{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0647bb0c",
   "metadata": {},
   "source": [
    "# Predicting antibody-antigen interactions with Transformer-based machine learning\n",
    "### Part 1: Data Clean & Dataset Generation\n",
    "This workbook contains all of the code for the data cleaning portion of the project\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c3179eae",
   "metadata": {},
   "source": [
    "### Step 0: Imports & Installations\n",
    "This project makes use of the BioPython package to assist in the reading, parsing, and extraction of information from FASTA sequence files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "08817a42",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Code to install biopython package; uncomment lines below to install the package if not already installed\n",
    "# import sys\n",
    "# !conda install --yes --prefix {sys.prefix} biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "66707ac8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd # pandas package\n",
    "import numpy as np # numpy package\n",
    "import re # for regex\n",
    "import os # os\n",
    "import urllib.request # urllib package for making GET requests\n",
    "from Bio import SeqIO # biopython package for parsing and reading FASTA files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31770e52",
   "metadata": {},
   "source": [
    "### Step 1: Read and import data\n",
    "We will be making use of the dataset from the Coronavirus Antibody Database by the Oxford Protein Informatics Group in this project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "fc630c12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Ab or Nb</th>\n",
       "      <th>Binds to</th>\n",
       "      <th>Doesn't Bind to</th>\n",
       "      <th>Neutralising Vs</th>\n",
       "      <th>Not Neutralising Vs</th>\n",
       "      <th>Protein + Epitope</th>\n",
       "      <th>Origin</th>\n",
       "      <th>VHorVHH</th>\n",
       "      <th>VL</th>\n",
       "      <th>...</th>\n",
       "      <th>Light J Gene</th>\n",
       "      <th>CDRH3</th>\n",
       "      <th>CDRL3</th>\n",
       "      <th>Structures</th>\n",
       "      <th>ABB Homology Model (if no structure)</th>\n",
       "      <th>Sources</th>\n",
       "      <th>Date Added</th>\n",
       "      <th>Last Updated</th>\n",
       "      <th>Update Description</th>\n",
       "      <th>Notes/Following Up?</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>029-2E1</td>\n",
       "      <td>Ab</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S; RBD</td>\n",
       "      <td>B-cells; SARS-CoV2_WT Human Patient</td>\n",
       "      <td>EVQLVESGGGLIQPGGSLRLSCAASGLTVSSNYMNWVRQAPGKGLE...</td>\n",
       "      <td>DIQMTQSPSSLSASVGDRVTITCQASQDISNYLNWYQQKPGKAPKL...</td>\n",
       "      <td>...</td>\n",
       "      <td>IGKJ5 (Human)</td>\n",
       "      <td>ARFRWGDV</td>\n",
       "      <td>QQYDNLPIT</td>\n",
       "      <td>ND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanjeev Kumar et al., 2022 (https://www.scienc...</td>\n",
       "      <td>Dec 20, 2022</td>\n",
       "      <td>Dec 20, 2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>029-1A4</td>\n",
       "      <td>Ab</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S; RBD</td>\n",
       "      <td>B-cells; SARS-CoV2_WT Human Patient</td>\n",
       "      <td>EVQLVQSGGGLVQPGGSLRLSCLASGLTFSSYEFNWIRQAPGKGLE...</td>\n",
       "      <td>QAGLTQPPSVSKGLRQTATLTCTGNSNDVGRQGAAWLQQHQGHPPK...</td>\n",
       "      <td>...</td>\n",
       "      <td>IGLJ1 (Human)</td>\n",
       "      <td>VTDLPGDLEFDF</td>\n",
       "      <td>SSWDSSRGGYV</td>\n",
       "      <td>ND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanjeev Kumar et al., 2022 (https://www.scienc...</td>\n",
       "      <td>Dec 20, 2022</td>\n",
       "      <td>Dec 20, 2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>029-1C1</td>\n",
       "      <td>Ab</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S; RBD</td>\n",
       "      <td>B-cells; SARS-CoV2_WT Human Patient</td>\n",
       "      <td>QVQLVQSGAEVKRPGASVKVLCMASGYSFTNYGINWVRQAPGQGLE...</td>\n",
       "      <td>QSVLTQPPSVAAAPGQKVTISWSGSSSNIGNNYVSWYQQVPGTAPK...</td>\n",
       "      <td>...</td>\n",
       "      <td>IGLJ2 (Human)</td>\n",
       "      <td>ARGSVLREADY</td>\n",
       "      <td>GTWDSSLNSLVV</td>\n",
       "      <td>ND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanjeev Kumar et al., 2022 (https://www.scienc...</td>\n",
       "      <td>Dec 20, 2022</td>\n",
       "      <td>Dec 20, 2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>029-1C3</td>\n",
       "      <td>Ab</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S; RBD</td>\n",
       "      <td>B-cells; SARS-CoV2_WT Human Patient</td>\n",
       "      <td>EVQLVQSGAEVKKPRESLKISCKGSGYNFTSYWIGWVRQMPGKGLE...</td>\n",
       "      <td>EIVLTQSPGTLSLSPGERATLSCRASQGFSSNSLAWYQQRPGQAPR...</td>\n",
       "      <td>...</td>\n",
       "      <td>IGKJ1 (Human)</td>\n",
       "      <td>SRPPYCSGTSCLDF</td>\n",
       "      <td>QQSHRSSMYT</td>\n",
       "      <td>ND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanjeev Kumar et al., 2022 (https://www.scienc...</td>\n",
       "      <td>Dec 20, 2022</td>\n",
       "      <td>Dec 20, 2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>029-1D1</td>\n",
       "      <td>Ab</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S; RBD</td>\n",
       "      <td>B-cells; SARS-CoV2_WT Human Patient</td>\n",
       "      <td>EVQLVESGGGLVQPGGSLRLSCAASRFTFANYWMSWVRQAPGKGLE...</td>\n",
       "      <td>DIQMTQSPSSLSASVGDRVTITCRASQAISNWLAWYQQKPEKAPKS...</td>\n",
       "      <td>...</td>\n",
       "      <td>IGKJ4 (Human)</td>\n",
       "      <td>ARGGPSRGTEVRPDDVFDM</td>\n",
       "      <td>QQYSSYPLT</td>\n",
       "      <td>ND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanjeev Kumar et al., 2022 (https://www.scienc...</td>\n",
       "      <td>Dec 20, 2022</td>\n",
       "      <td>Dec 20, 2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Name Ab or Nb      Binds to Doesn't Bind to Neutralising Vs  \\\n",
       "0  029-2E1       Ab  SARS-CoV2_WT             NaN             NaN   \n",
       "1  029-1A4       Ab  SARS-CoV2_WT             NaN             NaN   \n",
       "2  029-1C1       Ab  SARS-CoV2_WT             NaN             NaN   \n",
       "3  029-1C3       Ab  SARS-CoV2_WT             NaN             NaN   \n",
       "4  029-1D1       Ab  SARS-CoV2_WT             NaN             NaN   \n",
       "\n",
       "  Not Neutralising Vs Protein + Epitope                               Origin  \\\n",
       "0                 NaN            S; RBD  B-cells; SARS-CoV2_WT Human Patient   \n",
       "1                 NaN            S; RBD  B-cells; SARS-CoV2_WT Human Patient   \n",
       "2                 NaN            S; RBD  B-cells; SARS-CoV2_WT Human Patient   \n",
       "3                 NaN            S; RBD  B-cells; SARS-CoV2_WT Human Patient   \n",
       "4                 NaN            S; RBD  B-cells; SARS-CoV2_WT Human Patient   \n",
       "\n",
       "                                             VHorVHH  \\\n",
       "0  EVQLVESGGGLIQPGGSLRLSCAASGLTVSSNYMNWVRQAPGKGLE...   \n",
       "1  EVQLVQSGGGLVQPGGSLRLSCLASGLTFSSYEFNWIRQAPGKGLE...   \n",
       "2  QVQLVQSGAEVKRPGASVKVLCMASGYSFTNYGINWVRQAPGQGLE...   \n",
       "3  EVQLVQSGAEVKKPRESLKISCKGSGYNFTSYWIGWVRQMPGKGLE...   \n",
       "4  EVQLVESGGGLVQPGGSLRLSCAASRFTFANYWMSWVRQAPGKGLE...   \n",
       "\n",
       "                                                  VL  ...   Light J Gene  \\\n",
       "0  DIQMTQSPSSLSASVGDRVTITCQASQDISNYLNWYQQKPGKAPKL...  ...  IGKJ5 (Human)   \n",
       "1  QAGLTQPPSVSKGLRQTATLTCTGNSNDVGRQGAAWLQQHQGHPPK...  ...  IGLJ1 (Human)   \n",
       "2  QSVLTQPPSVAAAPGQKVTISWSGSSSNIGNNYVSWYQQVPGTAPK...  ...  IGLJ2 (Human)   \n",
       "3  EIVLTQSPGTLSLSPGERATLSCRASQGFSSNSLAWYQQRPGQAPR...  ...  IGKJ1 (Human)   \n",
       "4  DIQMTQSPSSLSASVGDRVTITCRASQAISNWLAWYQQKPEKAPKS...  ...  IGKJ4 (Human)   \n",
       "\n",
       "                 CDRH3         CDRL3 Structures  \\\n",
       "0             ARFRWGDV     QQYDNLPIT         ND   \n",
       "1         VTDLPGDLEFDF   SSWDSSRGGYV         ND   \n",
       "2          ARGSVLREADY  GTWDSSLNSLVV         ND   \n",
       "3       SRPPYCSGTSCLDF    QQSHRSSMYT         ND   \n",
       "4  ARGGPSRGTEVRPDDVFDM     QQYSSYPLT         ND   \n",
       "\n",
       "  ABB Homology Model (if no structure)  \\\n",
       "0                                  NaN   \n",
       "1                                  NaN   \n",
       "2                                  NaN   \n",
       "3                                  NaN   \n",
       "4                                  NaN   \n",
       "\n",
       "                                             Sources    Date Added  \\\n",
       "0  Sanjeev Kumar et al., 2022 (https://www.scienc...  Dec 20, 2022   \n",
       "1  Sanjeev Kumar et al., 2022 (https://www.scienc...  Dec 20, 2022   \n",
       "2  Sanjeev Kumar et al., 2022 (https://www.scienc...  Dec 20, 2022   \n",
       "3  Sanjeev Kumar et al., 2022 (https://www.scienc...  Dec 20, 2022   \n",
       "4  Sanjeev Kumar et al., 2022 (https://www.scienc...  Dec 20, 2022   \n",
       "\n",
       "   Last Updated Update Description Notes/Following Up?  \n",
       "0  Dec 20, 2022                NaN            Complete  \n",
       "1  Dec 20, 2022                NaN            Complete  \n",
       "2  Dec 20, 2022                NaN            Complete  \n",
       "3  Dec 20, 2022                NaN            Complete  \n",
       "4  Dec 20, 2022                NaN            Complete  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Read and import data\n",
    "covabdab_filepath = \"data/CoV-AbDab_201222.csv\"\n",
    "covabdab_df = pd.read_csv(covabdab_filepath)\n",
    "covabdab_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b08e9d41",
   "metadata": {},
   "source": [
    "We note that several series of data are present\n",
    "- The protein sequences for the Heavy & Light chains of the antibody\n",
    "- Structures that were referenced, and its sources\n",
    "- Information about interactions (binds to/doesn't bind to)\n",
    "- Information about neutralization (neutralizing vs/not neutralizing vs)\n",
    "\n",
    "There is unfortunately no information on the virus protein sequences other than the referenced structures"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "45de7e37",
   "metadata": {},
   "source": [
    "### Step 2: Obtain unique virus values\n",
    "- We note that each set of heavy and light chains may interact with or neutralize multiple virus types as well as their variants.\n",
    "- Here we obtain all the possible unique values for the virus types for mapping later"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "28ffe085",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to obtain unique from a column with delimited values\n",
    "def obtain_unique_values(df, columnName, delimiter):\n",
    "    result = [];\n",
    "    \n",
    "    columnUniqueValues = df[columnName].unique()\n",
    "    for value in columnUniqueValues:\n",
    "        result = result + str(value).split(delimiter)\n",
    "    \n",
    "    result = pd.Series(result)\n",
    "    result = result.unique()\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "89e1cf1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['SARS-CoV2_WT', 'SARS-CoV2_Beta', 'SARS-CoV2_Gamma',\n",
       "       'SARS-CoV2_ Beta', 'SARS-CoV2_Gamma (weak)', 'SARS-CoV2_Delta',\n",
       "       'SARS-CoV2_Alpha', 'SARS-CoV2_Omicron-BA1',\n",
       "       'SARS-CoV2_Omicron-BA2', 'SARS-CoV2_Kappa', 'SARS-CoV1',\n",
       "       'MERS-CoV', 'HKU1', 'OC43', 'SARS-CoV2_Epsilon', 'SARS-CoV2_Eta',\n",
       "       'SARS-CoV2_Zeta', 'SARS-CoV2_Iota', 'SARS-CoV2_Lambda',\n",
       "       'SARS-CoV2_Mu', 'SARS-CoV2_Omicron-BA3', 'SARS-CoV2_Omicron-BA4',\n",
       "       'SARS-CoV2_Omicron-BA5', 'SARS-CoV2_Omicron-BA2.12.1',\n",
       "       'SARS-CoV2_Omicron-BA2.75', 'SARS-CoV2_Omicron-BA2.38',\n",
       "       'SARS-CoV2_Omicron-BA2.76', '229E', 'SARS-COV2_Omicron-BA2.12.1',\n",
       "       'SARS-CoV2_Omicron-BA4/BA5', 'SARS-CoV2_Omicron-BA1 (weak)',\n",
       "       'SARS-CoV2_Omicron-BA2 (weak)', 'SARS-CoV2-Delta', 'Pangolin-GD',\n",
       "       'RaTG13', 'SARS-CoV2_Omicron-BA4/5', 'SARS_CoV2_Delta', 'RaTG013',\n",
       "       'Pangolin-GXP2V', 'A021', 'Civet007-2004', 'CS24', 'Frankfurt-1',\n",
       "       'LYRa11', 'Rs4231', 'SHC014', 'WIV1', 'HKU3', 'Rs4081',\n",
       "       'Rf1-2004 (weak)', 'SARS-CoV2_Omicron-BA1.1',\n",
       "       'SARS-CoV2_Omicron-BQ1', 'NL63', '299E', '339E', 'pang17',\n",
       "       'Rs7327', 'WIV-1', 'RmYN02', 'Yun11', 'Rf1', 'Rf4092', 'YN2013',\n",
       "       'ZC45', 'HKU3-1', 'Shaanxi2011', 'Rp3', 'BtKY72',\n",
       "       'SARS-CoV1 (weak)', 'Rs7327 (weak)', 'SHC014 (weak)', 'BM-4831',\n",
       "       'BM-4831 (weak)', 'YN2013 (weak)', 'Pangolin-GX',\n",
       "       'Pangolin-GX (weak)', 'RsSTT200', 'Pang17', 'LyRA3', 'C028',\n",
       "       'Khosta2', 'WIV-1 (weak)', 'LyRa3', 'RmYN02 (weak)', 'Rf1 (weak)',\n",
       "       'Rs4081 (weak)', 'BM4831', 'SARS-CoV2_WT (weak)',\n",
       "       'SARS-CoV2_Beta (weak)', 'SARS-CoV2_Omicron-BA2.13',\n",
       "       'SARS-CoV2_Omicron-BQ1.1', 'SARS-CoV2_Omicron-XBB',\n",
       "       'SARS-CoV2_Omicron-XBB1', 'SARS-CoV2_2.75',\n",
       "       'SARS-CoV2_Omicron-BA2.11', 'SARS-CoV2_Omicron-BA2.4',\n",
       "       'SARS-CoV2_Omicron-BA2.5', 'SARS-CoV2_Omicron-BA2.10.4',\n",
       "       'SARS-CoV2_Omcron-BA4', 'SARS-CoV1(weak)',\n",
       "       'SARS-CoV2_Omicron-BA1.1 (weak)', 'LyRa11', 'Rs4084', 'Yun2011',\n",
       "       'SARS-CoV2_Omicron-BA4.6', 'SARS-CoV2_Omicron-BA4.7',\n",
       "       'SARS-CoV2_Omicron-BA5.9', 'SARS-CoV2_Mu (weak)', 'Rs4231 (weak)',\n",
       "       'SARS-CoV2_Omicron-XBB.1', 'SARS_CoV2_Omicron-BQ1.1',\n",
       "       'SARS-CoV2_Omicron-BA2.75.2', 'NL63,229E,WIV-1', 'RsSHC014', 'HKU',\n",
       "       'PC4-127', 'Sin852', 'GZ-C', 'Urbani', 'BM48-31',\n",
       "       'SARS-CoV2_OmicronBA1', 'NL63, 229E, WIV-1', 'RatG13',\n",
       "       'SARS-CoV2_Omicron2.38', 'nan', 'SHAANXI-2011', 'RP3', 'SC2018',\n",
       "       'ZXC21', 'Rs4247', 'Anlong112', 'SARS-CoV2_Omicron-CA1',\n",
       "       'SARS-CoV2_Omicron-BR2', 'SARS-CoV2_Omicron-BM1.1.1', 'TGEV',\n",
       "       'HIV-1', 'Viral Nucleic Acid Chains', 'HKU1 (weak)',\n",
       "       'SRAS-CoV2_Beta', 'SARS-CoV2_Alpha (weak)', 'SARS-CoV2_Beta ',\n",
       "       'SARS-CoV2_Delta (weak)', 'RaTG13 (weak)', 'RC-o319',\n",
       "       'SARS_Omicron-BA1', 'HKU4', 'SARS-CoV2-Eta', 'SARS-CoV2-Alpha',\n",
       "       'SARS-CoV2_Omicron', 'SARS-COV2_Omicron-BQ1.1', 'SARS-CoV2-BA2.75',\n",
       "       'SARS-CoV2_WTSARS-CoV1', 'MERS', 'SARS_CoV2_Alpha', 'Bovine-CoV',\n",
       "       'PCoV GXP4L', 'BCoV RsSHC014', 'BCoV RaTG13',\n",
       "       'SARS-CoV2_BetaSARS-CoV2_Gamma', 'SARS-CoV2-BA3', 'GD1',\n",
       "       'bat-SL-CoV-WIV-1', 'PaGX', 'Viral Envelopes', 'Rs3367',\n",
       "       'SARS-CoV-1 (weak)', 'SARS-CoV-1', 'MERS-CoV (weak)',\n",
       "       'SARS-CoV_Beta', '229E (weak)', 'NL63 (weak)', 'OC43 (weak)',\n",
       "       'pang17 (weak)', 'SARS-CoV2_Omicron-2.75',\n",
       "       'SARS_CoV2_Omicron-BA4.7', 'SARS-CoV2_Epsilon (weak)', 'SARS-CoV',\n",
       "       'SARS-CoV (weak)', 'SARS_CoV2_Omicron-BA2', 'SARS-CoV2', 'WA-1',\n",
       "       'Civet 007-2004', 'SARS-CoV2_Omicron (weak)',\n",
       "       'SARS-CoV2_Kappa (weak)', 'SARS-CoV_Gamma (weak)'], dtype=object)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Obtain overall unique dictionary\n",
    "overallResult = obtain_unique_values(covabdab_df, \"Binds to\", \";\")\n",
    "np.append(overallResult, obtain_unique_values(covabdab_df, \"Doesn't Bind to\", \";\"))\n",
    "np.append(overallResult, obtain_unique_values(covabdab_df, \"Neutralising Vs\", \";\"))\n",
    "np.append(overallResult, obtain_unique_values(covabdab_df, \"Not Neutralising Vs\", \";\"))\n",
    "overallResult = pd.Series(overallResult).unique()\n",
    "overallResult\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e346fcc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.Series(overallResult).to_csv(\"virus_names.csv\", index=False, header=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "12c05991",
   "metadata": {},
   "source": [
    "### Step 3: Download structure data referenced in CoVAbDb\n",
    "We note that many of the records within the CoVAbDb dataset do not contain the virus protein sequences. However, they do contain the IDs of the referenced structures.\n",
    "\n",
    "These structures mostly contain:\n",
    "- Heavy and Light chains of one or more antibodies\n",
    "- The spike protein of the virus itself\n",
    "\n",
    "We will:\n",
    "- Download the FASTA sequence from RCSB Protein Data Bank (RCSB PDB) using their IDs\n",
    "- Using BioPython Library, we parse the FASTA Sequence (https://www.biostars.org/p/710/)\n",
    "- From each structure file, extract out the spike protein sequence to create a record for our training dataset\n",
    "- Save these outputs into a file: `fastaComplexes.csv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0b3d59f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out data that does not reference the structure files\n",
    "covabdab_df_with_structures = covabdab_df[covabdab_df['Structures'] != 'ND']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8b38c0e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function downloads the FASTA file from RCSB given its ID or opens the fasta file it is already downloaded and returns the sequence data\n",
    "def get_fasta_from_id(id):\n",
    "    # Normalize id name\n",
    "    id = id.upper()\n",
    "    fasta_file_path = \"fasta/\"+id+\".fasta\"\n",
    "\n",
    "    fasta_sequences = []\n",
    "\n",
    "    #if fasta file does not exist then download it from rcsdb\n",
    "    if not os.path.exists(path=\"fasta/\"+id+\".fasta\"):\n",
    "        url = \"https://www.rcsb.org/fasta/entry/\"+id+\"/display\"\n",
    "        try:\n",
    "            urllib.request.urlretrieve(url, fasta_file_path)\n",
    "        except:\n",
    "            print(\"Error occured while obtaining FASTA file for id: \"+id)\n",
    "            return fasta_sequences\n",
    "\n",
    "    #if file exists then open it directly\n",
    "    for seq_record in SeqIO.parse(fasta_file_path, \"fasta\"):\n",
    "        seq_prelim_virus_type, seq_prelim_protein_type = get_prelim_virus_type(seq_record.description)\n",
    "        fasta_sequences.append({\n",
    "            \"id\": seq_record.id,\n",
    "            \"prelim_virus\": seq_prelim_virus_type,\n",
    "            \"prelim_protein\": seq_prelim_protein_type,\n",
    "            \"description\": seq_record.description,\n",
    "            \"sequence\": str(seq_record.seq)\n",
    "        })\n",
    "    \n",
    "    return fasta_sequences\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0f02ebbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function assigns the preliminary virus type and protein type (Ab or Spike) based on the description of a FASTA sequence\n",
    "def get_prelim_virus_type(description):\n",
    "    normalzied_description = description.upper()\n",
    "    prelim_virus_type = \"\"\n",
    "    prelim_protein_type=\"\"\n",
    "    if \"coronavirus 2\".upper() in normalzied_description:\n",
    "        prelim_virus_type = \"SARS-CoV2\"\n",
    "    elif \"SARS coronavirus\".upper() in normalzied_description:\n",
    "        prelim_virus_type = \"SARS-CoV\"\n",
    "    elif \"Middle East\".upper() in normalzied_description:\n",
    "        prelim_virus_type = \"MERS-CoV\"\n",
    "    \n",
    "    if \"spike glycoprotein\".upper() in normalzied_description:\n",
    "        prelim_protein_type = \"Spike Glycoprotein\"\n",
    "    elif \"spike protein\".upper() in normalzied_description:\n",
    "        prelim_protein_type = \"Spike Protein\"\n",
    "    elif \"vl\".upper() in normalzied_description or \"light\".upper() in normalzied_description:\n",
    "        prelim_protein_type = \"Light Chain\"\n",
    "        prelim_virus_type = \"Antibody\"\n",
    "    elif \"vh\".upper() in normalzied_description or \"heavy\".upper() in normalzied_description:\n",
    "        prelim_protein_type = \"Heavy Chain\"\n",
    "        prelim_virus_type = \"Antibody\"\n",
    "    elif \"homo\".upper() in normalzied_description:\n",
    "        prelim_protein_type = \"Heavy/Light Chain\"\n",
    "        prelim_virus_type = \"Antibody\"\n",
    "    elif \"Nanobody\".upper() in normalzied_description:\n",
    "        prelim_protein_type = \"Nanobody\"\n",
    "        prelim_virus_type = \"Antibody\"\n",
    "        \n",
    "    return prelim_virus_type, prelim_protein_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d7d57f06",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error occured while obtaining FASTA file for id: 7N01\n",
      "Error occured while obtaining FASTA file for id: 7NY5\n",
      "Error occured while obtaining FASTA file for id: 7KGL\n",
      "Error occured while obtaining FASTA file for id: 7B11\n"
     ]
    }
   ],
   "source": [
    "combined_dataset = []\n",
    "\n",
    "# Iterate through the rows and construct new dataset\n",
    "for index, row in covabdab_df_with_structures.iterrows():\n",
    "    vhh_chain = str(row['VHorVHH'])\n",
    "    vl_chain = str(row['VL'])\n",
    "\n",
    "    # Use regex to extract out all the RCSB data\n",
    "    structure_ids = re.findall(\"[0-9A-Z]{4,}\",str(row['Structures']) )\n",
    "    for structure_id in structure_ids:\n",
    "        fasta_sequences = get_fasta_from_id(structure_id)\n",
    "\n",
    "        # Iterate through all the fasta sequences in the set\n",
    "        for fasta_sequence in fasta_sequences:\n",
    "            #Skip if this is one of our own vl or vhh chains\n",
    "            if vhh_chain in fasta_sequence['sequence']:\n",
    "                vhh_chain = fasta_sequence['sequence']\n",
    "                continue\n",
    "            elif vl_chain in fasta_sequence['sequence']:\n",
    "                vl_chain = fasta_sequence['sequence']\n",
    "                continue\n",
    "            \n",
    "            prelim_virus_type, prelim_protein_type = get_prelim_virus_type(fasta_sequence['description'])\n",
    "            if prelim_virus_type == \"Antibody\":\n",
    "                continue\n",
    "\n",
    "            combined_dataset.append({\n",
    "                \"name\": row['Name'],\n",
    "                \"structure_id\": structure_id,\n",
    "                \"vh\": vhh_chain,\n",
    "                \"vl\": vl_chain,\n",
    "                \"virus\": fasta_sequence['sequence'],\n",
    "                \"description\": fasta_sequence['description'],\n",
    "                \"binds_to\": row['Binds to'],\n",
    "                \"doesnt_bind_to\": row['Doesn\\'t Bind to'],\n",
    "                \"neutralising_vs\": row['Neutralising Vs'],\n",
    "                \"not_neutralising_vs\": row['Not Neutralising Vs'],\n",
    "            })\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2625b34a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>structure_id</th>\n",
       "      <th>vh</th>\n",
       "      <th>vl</th>\n",
       "      <th>virus</th>\n",
       "      <th>description</th>\n",
       "      <th>binds_to</th>\n",
       "      <th>doesnt_bind_to</th>\n",
       "      <th>neutralising_vs</th>\n",
       "      <th>not_neutralising_vs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2C7</td>\n",
       "      <td>7X2M</td>\n",
       "      <td>QVQLQESGGGLVQPGGSLRLSCAASGDTLDLYAIGWFRQTPGEERE...</td>\n",
       "      <td>nan</td>\n",
       "      <td>TNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFK...</td>\n",
       "      <td>7X2M_1|Chain A[auth E]|Spike protein S1|Severe...</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P2B4</td>\n",
       "      <td>8DXS</td>\n",
       "      <td>EVQLVESGGGLVQPGRSLRLSCAASGFNFDDYAMHWARQVPGKGLE...</td>\n",
       "      <td>DIQMTQSPSSLSASVGDRVTITCRASQSIHSFLSWYQQKPGKAPKL...</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>8DXS_1|Chains A, B, C|Spike glycoprotein|Sever...</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P1D9</td>\n",
       "      <td>8DWA</td>\n",
       "      <td>VQLVQSGAEVKKPGSSVRVSCKASGGTFGDDSITWVRQAPGQGLEW...</td>\n",
       "      <td>DIQMTQSPSSLSASVGDTVTITCRAGQTINTFLNWYQQKPGKAPKL...</td>\n",
       "      <td>LCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFKCY...</td>\n",
       "      <td>8DWA_1|Chain A|Spike protein S1|Severe acute r...</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D29</td>\n",
       "      <td>8DW9</td>\n",
       "      <td>QLVESGAEVKKPGSSVKVSCKASGDTFYTFDISWVRQAPGQGLEWM...</td>\n",
       "      <td>EIVLTQSPGTLSLSPGERATLSCRASQSVGNIYLAWYQQKPGQAPR...</td>\n",
       "      <td>TNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFK...</td>\n",
       "      <td>8DW9_1|Chains A, B|Spike protein S1|Severe acu...</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ab159</td>\n",
       "      <td>7X8Y</td>\n",
       "      <td>QVQLVESGGGLVKPGGSLRLSCAASGFTFNNAWMSWVRQAPGKGLE...</td>\n",
       "      <td>DIQMTQSPSSLSASVGDRVTITCRASQSITNYLNWYQQKPGKAPKF...</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>7X8Y_1|Chain A|Spike glycoprotein|Severe acute...</td>\n",
       "      <td>SARS-CoV2_WT;SARS-CoV2_Alpha;SARS-CoV2_Beta;SA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT;SARS-CoV2_Alpha;SARS-CoV2_Beta;SA...</td>\n",
       "      <td>SARS-CoV2_Delta;SARS-CoV2_Omicron-BA1;SARS-CoV...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    name structure_id                                                 vh  \\\n",
       "0  1-2C7         7X2M  QVQLQESGGGLVQPGGSLRLSCAASGDTLDLYAIGWFRQTPGEERE...   \n",
       "1   P2B4         8DXS  EVQLVESGGGLVQPGRSLRLSCAASGFNFDDYAMHWARQVPGKGLE...   \n",
       "2   P1D9         8DWA  VQLVQSGAEVKKPGSSVRVSCKASGGTFGDDSITWVRQAPGQGLEW...   \n",
       "3    D29         8DW9  QLVESGAEVKKPGSSVKVSCKASGDTFYTFDISWVRQAPGQGLEWM...   \n",
       "4  Ab159         7X8Y  QVQLVESGGGLVKPGGSLRLSCAASGFTFNNAWMSWVRQAPGKGLE...   \n",
       "\n",
       "                                                  vl  \\\n",
       "0                                                nan   \n",
       "1  DIQMTQSPSSLSASVGDRVTITCRASQSIHSFLSWYQQKPGKAPKL...   \n",
       "2  DIQMTQSPSSLSASVGDTVTITCRAGQTINTFLNWYQQKPGKAPKL...   \n",
       "3  EIVLTQSPGTLSLSPGERATLSCRASQSVGNIYLAWYQQKPGQAPR...   \n",
       "4  DIQMTQSPSSLSASVGDRVTITCRASQSITNYLNWYQQKPGKAPKF...   \n",
       "\n",
       "                                               virus  \\\n",
       "0  TNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFK...   \n",
       "1  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...   \n",
       "2  LCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFKCY...   \n",
       "3  TNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFK...   \n",
       "4  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...   \n",
       "\n",
       "                                         description  \\\n",
       "0  7X2M_1|Chain A[auth E]|Spike protein S1|Severe...   \n",
       "1  8DXS_1|Chains A, B, C|Spike glycoprotein|Sever...   \n",
       "2  8DWA_1|Chain A|Spike protein S1|Severe acute r...   \n",
       "3  8DW9_1|Chains A, B|Spike protein S1|Severe acu...   \n",
       "4  7X8Y_1|Chain A|Spike glycoprotein|Severe acute...   \n",
       "\n",
       "                                            binds_to doesnt_bind_to  \\\n",
       "0                                       SARS-CoV2_WT            NaN   \n",
       "1                                       SARS-CoV2_WT            NaN   \n",
       "2                                       SARS-CoV2_WT            NaN   \n",
       "3                                       SARS-CoV2_WT            NaN   \n",
       "4  SARS-CoV2_WT;SARS-CoV2_Alpha;SARS-CoV2_Beta;SA...            NaN   \n",
       "\n",
       "                                     neutralising_vs  \\\n",
       "0                                       SARS-CoV2_WT   \n",
       "1                                       SARS-CoV2_WT   \n",
       "2                                       SARS-CoV2_WT   \n",
       "3                                       SARS-CoV2_WT   \n",
       "4  SARS-CoV2_WT;SARS-CoV2_Alpha;SARS-CoV2_Beta;SA...   \n",
       "\n",
       "                                 not_neutralising_vs  \n",
       "0                                                NaN  \n",
       "1                                                NaN  \n",
       "2                                                NaN  \n",
       "3                                                NaN  \n",
       "4  SARS-CoV2_Delta;SARS-CoV2_Omicron-BA1;SARS-CoV...  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create dataframe from it\n",
    "fasta_complexes_df = pd.DataFrame.from_records(combined_dataset)\n",
    "fasta_complexes_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ff423c3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_complexes_df.to_csv(\"fastaComplexes.csv\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "313f7304",
   "metadata": {},
   "source": [
    "A problem which we encounter here is that although we have the virus sequences, we do not know variants of which the belong to.\n",
    "\n",
    "To resolve this, we reference the workflow that was done in the Jupytope paper where we put these sequences through an application in GISAID called AudacityInstant, which finds the closest lineage of each sequence; and hence its variant.\n",
    "\n",
    "As AudacityInstant is only able to accept nucleotide sequences, we will have to translate these protein sequences using an application in National Center for Biotechnology Information called tBLASTn\n",
    "\n",
    "tBLASTn finds potential translated nucleotide sequences given a amino acid sequence which is what our protein sequences are encoded in."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08c4a928",
   "metadata": {},
   "source": [
    "### Step 4: Prepare data for tBLASTn\n",
    "Removal of duplicate values\n",
    "- Upon further investigation of the viral sequences, we note that many of the virus spike protein sequences are shared across multiple complexes.\n",
    "- To reduce the amount of searches required for tBLASTn, we extract out these unique virus protein sequences\n",
    "- We will keep the descriptions for the first instance of duplucate virus sequences\n",
    "\n",
    "Preparation of data for tBLASTn\n",
    "- We will perform tBLASTn for the sequences in batches of 10 due to NCBI's CPU usage limitations, as well as space constraints preventing us to perform a local BLAST search using the BLAST+ binaries\n",
    "- Spike protein sequences are split into groups of 10 and saved into individual FASTA files for upload and processing on the NCBI website"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "00db946f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>structure_id</th>\n",
       "      <th>vh</th>\n",
       "      <th>vl</th>\n",
       "      <th>virus</th>\n",
       "      <th>description</th>\n",
       "      <th>binds_to</th>\n",
       "      <th>doesnt_bind_to</th>\n",
       "      <th>neutralising_vs</th>\n",
       "      <th>not_neutralising_vs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2C7</td>\n",
       "      <td>7X2M</td>\n",
       "      <td>QVQLQESGGGLVQPGGSLRLSCAASGDTLDLYAIGWFRQTPGEERE...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFK...</td>\n",
       "      <td>7X2M_1|Chain A[auth E]|Spike protein S1|Severe...</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P2B4</td>\n",
       "      <td>8DXS</td>\n",
       "      <td>EVQLVESGGGLVQPGRSLRLSCAASGFNFDDYAMHWARQVPGKGLE...</td>\n",
       "      <td>DIQMTQSPSSLSASVGDRVTITCRASQSIHSFLSWYQQKPGKAPKL...</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>8DXS_1|Chains A, B, C|Spike glycoprotein|Sever...</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P1D9</td>\n",
       "      <td>8DWA</td>\n",
       "      <td>VQLVQSGAEVKKPGSSVRVSCKASGGTFGDDSITWVRQAPGQGLEW...</td>\n",
       "      <td>DIQMTQSPSSLSASVGDTVTITCRAGQTINTFLNWYQQKPGKAPKL...</td>\n",
       "      <td>LCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFKCY...</td>\n",
       "      <td>8DWA_1|Chain A|Spike protein S1|Severe acute r...</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D29</td>\n",
       "      <td>8DW9</td>\n",
       "      <td>QLVESGAEVKKPGSSVKVSCKASGDTFYTFDISWVRQAPGQGLEWM...</td>\n",
       "      <td>EIVLTQSPGTLSLSPGERATLSCRASQSVGNIYLAWYQQKPGQAPR...</td>\n",
       "      <td>TNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFK...</td>\n",
       "      <td>8DW9_1|Chains A, B|Spike protein S1|Severe acu...</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ab159</td>\n",
       "      <td>7X8Y</td>\n",
       "      <td>QVQLVESGGGLVKPGGSLRLSCAASGFTFNNAWMSWVRQAPGKGLE...</td>\n",
       "      <td>DIQMTQSPSSLSASVGDRVTITCRASQSITNYLNWYQQKPGKAPKF...</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>7X8Y_1|Chain A|Spike glycoprotein|Severe acute...</td>\n",
       "      <td>SARS-CoV2_WT;SARS-CoV2_Alpha;SARS-CoV2_Beta;SA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT;SARS-CoV2_Alpha;SARS-CoV2_Beta;SA...</td>\n",
       "      <td>SARS-CoV2_Delta;SARS-CoV2_Omicron-BA1;SARS-CoV...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    name structure_id                                                 vh  \\\n",
       "0  1-2C7         7X2M  QVQLQESGGGLVQPGGSLRLSCAASGDTLDLYAIGWFRQTPGEERE...   \n",
       "1   P2B4         8DXS  EVQLVESGGGLVQPGRSLRLSCAASGFNFDDYAMHWARQVPGKGLE...   \n",
       "2   P1D9         8DWA  VQLVQSGAEVKKPGSSVRVSCKASGGTFGDDSITWVRQAPGQGLEW...   \n",
       "3    D29         8DW9  QLVESGAEVKKPGSSVKVSCKASGDTFYTFDISWVRQAPGQGLEWM...   \n",
       "4  Ab159         7X8Y  QVQLVESGGGLVKPGGSLRLSCAASGFTFNNAWMSWVRQAPGKGLE...   \n",
       "\n",
       "                                                  vl  \\\n",
       "0                                                NaN   \n",
       "1  DIQMTQSPSSLSASVGDRVTITCRASQSIHSFLSWYQQKPGKAPKL...   \n",
       "2  DIQMTQSPSSLSASVGDTVTITCRAGQTINTFLNWYQQKPGKAPKL...   \n",
       "3  EIVLTQSPGTLSLSPGERATLSCRASQSVGNIYLAWYQQKPGQAPR...   \n",
       "4  DIQMTQSPSSLSASVGDRVTITCRASQSITNYLNWYQQKPGKAPKF...   \n",
       "\n",
       "                                               virus  \\\n",
       "0  TNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFK...   \n",
       "1  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...   \n",
       "2  LCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFKCY...   \n",
       "3  TNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFK...   \n",
       "4  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...   \n",
       "\n",
       "                                         description  \\\n",
       "0  7X2M_1|Chain A[auth E]|Spike protein S1|Severe...   \n",
       "1  8DXS_1|Chains A, B, C|Spike glycoprotein|Sever...   \n",
       "2  8DWA_1|Chain A|Spike protein S1|Severe acute r...   \n",
       "3  8DW9_1|Chains A, B|Spike protein S1|Severe acu...   \n",
       "4  7X8Y_1|Chain A|Spike glycoprotein|Severe acute...   \n",
       "\n",
       "                                            binds_to doesnt_bind_to  \\\n",
       "0                                       SARS-CoV2_WT            NaN   \n",
       "1                                       SARS-CoV2_WT            NaN   \n",
       "2                                       SARS-CoV2_WT            NaN   \n",
       "3                                       SARS-CoV2_WT            NaN   \n",
       "4  SARS-CoV2_WT;SARS-CoV2_Alpha;SARS-CoV2_Beta;SA...            NaN   \n",
       "\n",
       "                                     neutralising_vs  \\\n",
       "0                                       SARS-CoV2_WT   \n",
       "1                                       SARS-CoV2_WT   \n",
       "2                                       SARS-CoV2_WT   \n",
       "3                                       SARS-CoV2_WT   \n",
       "4  SARS-CoV2_WT;SARS-CoV2_Alpha;SARS-CoV2_Beta;SA...   \n",
       "\n",
       "                                 not_neutralising_vs  \n",
       "0                                                NaN  \n",
       "1                                                NaN  \n",
       "2                                                NaN  \n",
       "3                                                NaN  \n",
       "4  SARS-CoV2_Delta;SARS-CoV2_Omicron-BA1;SARS-CoV...  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_complexes_df = pd.read_csv(\"fastaComplexes.csv\")\n",
    "fasta_complexes_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "fdc6235e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(335, 10)\n"
     ]
    }
   ],
   "source": [
    "# Obtain unique values for the virus protein sequences\n",
    "unique_virus_protein_sequences = fasta_complexes_df.drop_duplicates('virus')\n",
    "print(unique_virus_protein_sequences.shape)\n",
    "unique_virus_protein_sequences.to_csv(\"fastaComplexesUNIQUE.csv\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "0f75efbd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(273, 10)\n"
     ]
    }
   ],
   "source": [
    "# Filter out sequences that are COVID-19 (contains 'coronavirus 2') only\n",
    "covid19_protein_sequences = unique_virus_protein_sequences[unique_virus_protein_sequences['description'].str.contains('coronavirus 2',case=False)]\n",
    "print(covid19_protein_sequences.shape)\n",
    "covid19_protein_sequences.to_csv(\"fastaComplexesUNIQUE_covid19.csv\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a9e49371",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate large FASTA file containing all sequences\n",
    "output = \"\"\n",
    "counter = 1\n",
    "fastaFile = open(\"covid19_sequences.fasta\",\"w\")\n",
    "for index, row in viral_protein_sequences_covid19.iterrows():\n",
    "    fastaFile.write(\">\" + row['description'] +\"\\n\" + row['virus'] + \"\\n\")\n",
    "fastaFile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "ab31df0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate batches of 10 sequences files\n",
    "counter = 0\n",
    "fastaFile = open(\"batches/covid19_sequences(\"+str(counter)+\").fasta\",\"a\")\n",
    "for index, row in viral_protein_sequences_covid19.iterrows():\n",
    "    if counter%10 == 0:\n",
    "        fastaFile.close()\n",
    "        fastaFile = open(\"batches/covid19_sequences(\"+str(counter)+\").fasta\",\"a\")\n",
    "    fastaFile.write(\">\" + row['description'] +\"\\n\" + row['virus'] + \"\\n\")\n",
    "    counter += 1\n",
    "fastaFile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5911ac41",
   "metadata": {},
   "source": [
    "These batches were put through the NCBI tBLASTn application searching under the `Betacoronavirus` database using the default parameters except for the `Max target sequences` which was adjusted to 10.\n",
    "\n",
    "The top or top 10 sequence results were downloaded every protein sequence that was put through tBLASTn. They are saved in the folder `tblastn-results` under the nomenclature `<PROTEIN SEQ ID>.txt`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "770efed1",
   "metadata": {},
   "source": [
    "### Step 5: tBLASTn Results & preparation for lineage classification in GISAID's AudacityInstant\n",
    "- After performing tBLASTn on the sequences we import all of them into a pandas df\n",
    "- tBLASTn's results are sorted by e-value which refers to the chances a sequence like that can be found in the database at random; hence, a lower value indicates a higher accuracy\n",
    "- We will take the top result for every protein sequence's tBLASTn results\n",
    "- This will be used to form a new excel spreadsheet for GSAID's AudacityInstant to determine the lineage/variant of the virus spike proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "693bde05",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>id_tblastn</th>\n",
       "      <th>sequence</th>\n",
       "      <th>nucleotide_desc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6WPS_1</td>\n",
       "      <td>OA965538.1</td>\n",
       "      <td>AGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTGGCTGTCACTC...</td>\n",
       "      <td>OA965538.1 Severe acute respiratory syndrome c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6XDG_1</td>\n",
       "      <td>OL454750.1</td>\n",
       "      <td>TACGTTGAAATCCTTCACTGTAGAAAAAGGAATCTATCAAACTTCT...</td>\n",
       "      <td>OL454750.1 Severe acute respiratory syndrome c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6XE1_3</td>\n",
       "      <td>OW680892.1</td>\n",
       "      <td>AGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTGGCTGTCACTC...</td>\n",
       "      <td>OW680892.1 Severe acute respiratory syndrome c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6XKP_1</td>\n",
       "      <td>OL454750.1</td>\n",
       "      <td>TACGTTGAAATCCTTCACTGTAGAAAAAGGAATCTATCAAACTTCT...</td>\n",
       "      <td>OL454750.1 Severe acute respiratory syndrome c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6YLA_1</td>\n",
       "      <td>OL444812.1</td>\n",
       "      <td>GTAGAAAAAGGAATCTATCAAACTTCTAACTTTAGAGTCCAACCAA...</td>\n",
       "      <td>OL444812.1 Severe acute respiratory syndrome c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>8DXS_1</td>\n",
       "      <td>OY602403.3</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...</td>\n",
       "      <td>OY602403.3 Severe acute respiratory syndrome c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>8DZH_1</td>\n",
       "      <td>ON334068.1</td>\n",
       "      <td>TAATAAAGGAGCTGGTGGCCATAGTTACGGCGCCGATCTAAAGTCA...</td>\n",
       "      <td>ON334068.1 Severe acute respiratory syndrome c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>8ERQ_3</td>\n",
       "      <td>ON334068.1</td>\n",
       "      <td>TAATAAAGGAGCTGGTGGCCATAGTTACGGCGCCGATCTAAAGTCA...</td>\n",
       "      <td>ON334068.1 Severe acute respiratory syndrome c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>8GX9_1</td>\n",
       "      <td>OL444814.1</td>\n",
       "      <td>GTAGAAAAAGGAATCTATCAAACTTCTAACTTTAGAGTCCAACCAA...</td>\n",
       "      <td>OL444814.1 Severe acute respiratory syndrome c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>8GZ5_1</td>\n",
       "      <td>OP964909.1</td>\n",
       "      <td>GTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACT...</td>\n",
       "      <td>OP964909.1 Severe acute respiratory syndrome c...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>260 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         id  id_tblastn                                           sequence  \\\n",
       "0    6WPS_1  OA965538.1  AGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTGGCTGTCACTC...   \n",
       "1    6XDG_1  OL454750.1  TACGTTGAAATCCTTCACTGTAGAAAAAGGAATCTATCAAACTTCT...   \n",
       "2    6XE1_3  OW680892.1  AGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTGGCTGTCACTC...   \n",
       "3    6XKP_1  OL454750.1  TACGTTGAAATCCTTCACTGTAGAAAAAGGAATCTATCAAACTTCT...   \n",
       "4    6YLA_1  OL444812.1  GTAGAAAAAGGAATCTATCAAACTTCTAACTTTAGAGTCCAACCAA...   \n",
       "..      ...         ...                                                ...   \n",
       "255  8DXS_1  OY602403.3  NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...   \n",
       "256  8DZH_1  ON334068.1  TAATAAAGGAGCTGGTGGCCATAGTTACGGCGCCGATCTAAAGTCA...   \n",
       "257  8ERQ_3  ON334068.1  TAATAAAGGAGCTGGTGGCCATAGTTACGGCGCCGATCTAAAGTCA...   \n",
       "258  8GX9_1  OL444814.1  GTAGAAAAAGGAATCTATCAAACTTCTAACTTTAGAGTCCAACCAA...   \n",
       "259  8GZ5_1  OP964909.1  GTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACT...   \n",
       "\n",
       "                                       nucleotide_desc  \n",
       "0    OA965538.1 Severe acute respiratory syndrome c...  \n",
       "1    OL454750.1 Severe acute respiratory syndrome c...  \n",
       "2    OW680892.1 Severe acute respiratory syndrome c...  \n",
       "3    OL454750.1 Severe acute respiratory syndrome c...  \n",
       "4    OL444812.1 Severe acute respiratory syndrome c...  \n",
       "..                                                 ...  \n",
       "255  OY602403.3 Severe acute respiratory syndrome c...  \n",
       "256  ON334068.1 Severe acute respiratory syndrome c...  \n",
       "257  ON334068.1 Severe acute respiratory syndrome c...  \n",
       "258  OL444814.1 Severe acute respiratory syndrome c...  \n",
       "259  OP964909.1 Severe acute respiratory syndrome c...  \n",
       "\n",
       "[260 rows x 4 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Read all fasta files and form an array of dictionaries\n",
    "tblastn_sequences = []\n",
    "\n",
    "for nucleotide_fasta in os.listdir('tblastn-results'):\n",
    "    for nucleotide_sequence in SeqIO.parse('tblastn-results/'+nucleotide_fasta,\"fasta\"):\n",
    "        tblastn_sequences.append({\n",
    "            \"id\":nucleotide_fasta.replace(\".txt\",\"\"),\n",
    "            \"id_tblastn\": nucleotide_sequence.id,\n",
    "            \"sequence\": str(nucleotide_sequence.seq),\n",
    "            \"nucleotide_desc\": nucleotide_sequence.description\n",
    "        })\n",
    "        break # because we take the top result only\n",
    "\n",
    "# Create a pandas dataframe from it\n",
    "tblastn_df = pd.DataFrame.from_records(tblastn_sequences)\n",
    "tblastn_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "09b322cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import previous unique fasta files\n",
    "fasta_sequences_covid19_df = pd.read_csv(\"fastaComplexesUNIQUE_covid19.csv\")\n",
    "fasta_sequences_covid19_df['virus_id'] =  fasta_sequences_covid19_df['description'].str[:6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "79bcc248",
   "metadata": {},
   "outputs": [],
   "source": [
    "#perform a union join between the 2 dataframes\n",
    "merged_tblastn_sequences_df = pd.merge(fasta_sequences_covid19_df,tblastn_df,how=\"outer\",left_on=\"virus_id\",right_on=\"id\")\n",
    "merged_tblastn_sequences_df\n",
    "\n",
    "merged_tblastn_sequences_df.to_csv(\"translated_sequences.csv\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0de9e3f",
   "metadata": {},
   "source": [
    "We note that there are a lot of duplicate top results for nucleotide translations; hence we will drop duplicates so that GISAID has a smaller set of records to work with.\n",
    "- Note that AudacityInstant has a base size limit of 250 to 40,000\n",
    "- We will remove records that have <250 bases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "44c9eccc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(80, 4)\n",
      "(78, 4)\n"
     ]
    }
   ],
   "source": [
    "tblastn_df_unique = tblastn_df.drop_duplicates('id_tblastn')\n",
    "print(tblastn_df_unique.shape)\n",
    "tblastn_df_unique = tblastn_df_unique[tblastn_df_unique['sequence'].str.len() >= 250]\n",
    "print(tblastn_df_unique.shape)\n",
    "tblastn_df_unique.to_csv(\"translated_sequences_unique.csv\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "bff3ea01",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add them all into a new fasta file\n",
    "nucleotide_fasta = open(\"nucleotide_sequences.fasta\",\"w\")\n",
    "for index, row in tblastn_df_unique.iterrows():\n",
    "    nucleotide_fasta.write(\">\" + row['nucleotide_desc'] +\"\\n\" + row['sequence'] + \"\\n\")\n",
    "nucleotide_fasta.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5616412a",
   "metadata": {},
   "source": [
    "Since the AudacityInsant application is on a web platform without an official API, the nucleotide sequences were put into the platform and the results were manually entered into the excel sheet."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55d521e4",
   "metadata": {},
   "source": [
    "### Step 6: Matching variants from nucleotide sequences to protein sequences\n",
    "Now that we have identified the variants of several nucleotide sequences, we can map them back to their pre-translated protein sequences\n",
    "\n",
    "We map the protein sequences to their original pre-translated protein sequences based on their PDB Id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "21e20bbf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(273, 15)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>structure_id</th>\n",
       "      <th>vh</th>\n",
       "      <th>vl</th>\n",
       "      <th>virus</th>\n",
       "      <th>description</th>\n",
       "      <th>binds_to</th>\n",
       "      <th>doesnt_bind_to</th>\n",
       "      <th>neutralising_vs</th>\n",
       "      <th>not_neutralising_vs</th>\n",
       "      <th>virus_id</th>\n",
       "      <th>id</th>\n",
       "      <th>id_tblastn</th>\n",
       "      <th>sequence</th>\n",
       "      <th>nucleotide_desc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-2C7</td>\n",
       "      <td>7X2M</td>\n",
       "      <td>QVQLQESGGGLVQPGGSLRLSCAASGDTLDLYAIGWFRQTPGEERE...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFK...</td>\n",
       "      <td>7X2M_1|Chain A[auth E]|Spike protein S1|Severe...</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7X2M_1</td>\n",
       "      <td>7X2M_1</td>\n",
       "      <td>OP964904.1</td>\n",
       "      <td>GTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACT...</td>\n",
       "      <td>OP964904.1 Severe acute respiratory syndrome c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P2B4</td>\n",
       "      <td>8DXS</td>\n",
       "      <td>EVQLVESGGGLVQPGRSLRLSCAASGFNFDDYAMHWARQVPGKGLE...</td>\n",
       "      <td>DIQMTQSPSSLSASVGDRVTITCRASQSIHSFLSWYQQKPGKAPKL...</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>8DXS_1|Chains A, B, C|Spike glycoprotein|Sever...</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8DXS_1</td>\n",
       "      <td>8DXS_1</td>\n",
       "      <td>OY602403.3</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...</td>\n",
       "      <td>OY602403.3 Severe acute respiratory syndrome c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P1D9</td>\n",
       "      <td>8DWA</td>\n",
       "      <td>VQLVQSGAEVKKPGSSVRVSCKASGGTFGDDSITWVRQAPGQGLEW...</td>\n",
       "      <td>DIQMTQSPSSLSASVGDTVTITCRAGQTINTFLNWYQQKPGKAPKL...</td>\n",
       "      <td>LCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFKCY...</td>\n",
       "      <td>8DWA_1|Chain A|Spike protein S1|Severe acute r...</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8DWA_1</td>\n",
       "      <td>8DWA_1</td>\n",
       "      <td>MZ242023.1</td>\n",
       "      <td>AAACTTGTGCCCTTTTGGTGAAGTTTTTAACGCCACCAGATTTGCT...</td>\n",
       "      <td>MZ242023.1 Severe acute respiratory syndrome c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D29</td>\n",
       "      <td>8DW9</td>\n",
       "      <td>QLVESGAEVKKPGSSVKVSCKASGDTFYTFDISWVRQAPGQGLEWM...</td>\n",
       "      <td>EIVLTQSPGTLSLSPGERATLSCRASQSVGNIYLAWYQQKPGQAPR...</td>\n",
       "      <td>TNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFK...</td>\n",
       "      <td>8DW9_1|Chains A, B|Spike protein S1|Severe acu...</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8DW9_1</td>\n",
       "      <td>8DW9_1</td>\n",
       "      <td>OL444814.1</td>\n",
       "      <td>GTAGAAAAAGGAATCTATCAAACTTCTAACTTTAGAGTCCAACCAA...</td>\n",
       "      <td>OL444814.1 Severe acute respiratory syndrome c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ab159</td>\n",
       "      <td>7X8Y</td>\n",
       "      <td>QVQLVESGGGLVKPGGSLRLSCAASGFTFNNAWMSWVRQAPGKGLE...</td>\n",
       "      <td>DIQMTQSPSSLSASVGDRVTITCRASQSITNYLNWYQQKPGKAPKF...</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>7X8Y_1|Chain A|Spike glycoprotein|Severe acute...</td>\n",
       "      <td>SARS-CoV2_WT;SARS-CoV2_Alpha;SARS-CoV2_Beta;SA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT;SARS-CoV2_Alpha;SARS-CoV2_Beta;SA...</td>\n",
       "      <td>SARS-CoV2_Delta;SARS-CoV2_Omicron-BA1;SARS-CoV...</td>\n",
       "      <td>7X8Y_1</td>\n",
       "      <td>7X8Y_1</td>\n",
       "      <td>OA965538.1</td>\n",
       "      <td>AGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTGGCTGTCACTC...</td>\n",
       "      <td>OA965538.1 Severe acute respiratory syndrome c...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    name structure_id                                                 vh  \\\n",
       "0  1-2C7         7X2M  QVQLQESGGGLVQPGGSLRLSCAASGDTLDLYAIGWFRQTPGEERE...   \n",
       "1   P2B4         8DXS  EVQLVESGGGLVQPGRSLRLSCAASGFNFDDYAMHWARQVPGKGLE...   \n",
       "2   P1D9         8DWA  VQLVQSGAEVKKPGSSVRVSCKASGGTFGDDSITWVRQAPGQGLEW...   \n",
       "3    D29         8DW9  QLVESGAEVKKPGSSVKVSCKASGDTFYTFDISWVRQAPGQGLEWM...   \n",
       "4  Ab159         7X8Y  QVQLVESGGGLVKPGGSLRLSCAASGFTFNNAWMSWVRQAPGKGLE...   \n",
       "\n",
       "                                                  vl  \\\n",
       "0                                                NaN   \n",
       "1  DIQMTQSPSSLSASVGDRVTITCRASQSIHSFLSWYQQKPGKAPKL...   \n",
       "2  DIQMTQSPSSLSASVGDTVTITCRAGQTINTFLNWYQQKPGKAPKL...   \n",
       "3  EIVLTQSPGTLSLSPGERATLSCRASQSVGNIYLAWYQQKPGQAPR...   \n",
       "4  DIQMTQSPSSLSASVGDRVTITCRASQSITNYLNWYQQKPGKAPKF...   \n",
       "\n",
       "                                               virus  \\\n",
       "0  TNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFK...   \n",
       "1  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...   \n",
       "2  LCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFKCY...   \n",
       "3  TNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFK...   \n",
       "4  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...   \n",
       "\n",
       "                                         description  \\\n",
       "0  7X2M_1|Chain A[auth E]|Spike protein S1|Severe...   \n",
       "1  8DXS_1|Chains A, B, C|Spike glycoprotein|Sever...   \n",
       "2  8DWA_1|Chain A|Spike protein S1|Severe acute r...   \n",
       "3  8DW9_1|Chains A, B|Spike protein S1|Severe acu...   \n",
       "4  7X8Y_1|Chain A|Spike glycoprotein|Severe acute...   \n",
       "\n",
       "                                            binds_to doesnt_bind_to  \\\n",
       "0                                       SARS-CoV2_WT            NaN   \n",
       "1                                       SARS-CoV2_WT            NaN   \n",
       "2                                       SARS-CoV2_WT            NaN   \n",
       "3                                       SARS-CoV2_WT            NaN   \n",
       "4  SARS-CoV2_WT;SARS-CoV2_Alpha;SARS-CoV2_Beta;SA...            NaN   \n",
       "\n",
       "                                     neutralising_vs  \\\n",
       "0                                       SARS-CoV2_WT   \n",
       "1                                       SARS-CoV2_WT   \n",
       "2                                       SARS-CoV2_WT   \n",
       "3                                       SARS-CoV2_WT   \n",
       "4  SARS-CoV2_WT;SARS-CoV2_Alpha;SARS-CoV2_Beta;SA...   \n",
       "\n",
       "                                 not_neutralising_vs virus_id      id  \\\n",
       "0                                                NaN   7X2M_1  7X2M_1   \n",
       "1                                                NaN   8DXS_1  8DXS_1   \n",
       "2                                                NaN   8DWA_1  8DWA_1   \n",
       "3                                                NaN   8DW9_1  8DW9_1   \n",
       "4  SARS-CoV2_Delta;SARS-CoV2_Omicron-BA1;SARS-CoV...   7X8Y_1  7X8Y_1   \n",
       "\n",
       "   id_tblastn                                           sequence  \\\n",
       "0  OP964904.1  GTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACT...   \n",
       "1  OY602403.3  NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...   \n",
       "2  MZ242023.1  AAACTTGTGCCCTTTTGGTGAAGTTTTTAACGCCACCAGATTTGCT...   \n",
       "3  OL444814.1  GTAGAAAAAGGAATCTATCAAACTTCTAACTTTAGAGTCCAACCAA...   \n",
       "4  OA965538.1  AGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTGGCTGTCACTC...   \n",
       "\n",
       "                                     nucleotide_desc  \n",
       "0  OP964904.1 Severe acute respiratory syndrome c...  \n",
       "1  OY602403.3 Severe acute respiratory syndrome c...  \n",
       "2  MZ242023.1 Severe acute respiratory syndrome c...  \n",
       "3  OL444814.1 Severe acute respiratory syndrome c...  \n",
       "4  OA965538.1 Severe acute respiratory syndrome c...  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Retrieve translated mapped protein sequences from earlier\n",
    "translated_virus_sequences = pd.read_csv(\"translated_sequences.csv\")\n",
    "print(translated_virus_sequences.shape)\n",
    "translated_virus_sequences.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ad07cce4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(78, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>id_tblastn</th>\n",
       "      <th>sequence</th>\n",
       "      <th>nucleotide_desc</th>\n",
       "      <th>lineage</th>\n",
       "      <th>who_name</th>\n",
       "      <th>comment</th>\n",
       "      <th>mapped_virus_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6WPS_1</td>\n",
       "      <td>OA965538.1</td>\n",
       "      <td>AGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTGGCTGTCACTC...</td>\n",
       "      <td>OA965538.1 Severe acute respiratory syndrome c...</td>\n",
       "      <td>B.39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6XDG_1</td>\n",
       "      <td>OL454750.1</td>\n",
       "      <td>TACGTTGAAATCCTTCACTGTAGAAAAAGGAATCTATCAAACTTCT...</td>\n",
       "      <td>OL454750.1 Severe acute respiratory syndrome c...</td>\n",
       "      <td>B.1.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6XE1_3</td>\n",
       "      <td>OW680892.1</td>\n",
       "      <td>AGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTGGCTGTCACTC...</td>\n",
       "      <td>OW680892.1 Severe acute respiratory syndrome c...</td>\n",
       "      <td>B.1.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6YLA_1</td>\n",
       "      <td>OL444812.1</td>\n",
       "      <td>GTAGAAAAAGGAATCTATCAAACTTCTAACTTTAGAGTCCAACCAA...</td>\n",
       "      <td>OL444812.1 Severe acute respiratory syndrome c...</td>\n",
       "      <td>B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6ZDG_1</td>\n",
       "      <td>OL444814.1</td>\n",
       "      <td>GTAGAAAAAGGAATCTATCAAACTTCTAACTTTAGAGTCCAACCAA...</td>\n",
       "      <td>OL444814.1 Severe acute respiratory syndrome c...</td>\n",
       "      <td>B.1.177.12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id  id_tblastn                                           sequence  \\\n",
       "0  6WPS_1  OA965538.1  AGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTGGCTGTCACTC...   \n",
       "1  6XDG_1  OL454750.1  TACGTTGAAATCCTTCACTGTAGAAAAAGGAATCTATCAAACTTCT...   \n",
       "2  6XE1_3  OW680892.1  AGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTGGCTGTCACTC...   \n",
       "3  6YLA_1  OL444812.1  GTAGAAAAAGGAATCTATCAAACTTCTAACTTTAGAGTCCAACCAA...   \n",
       "4  6ZDG_1  OL444814.1  GTAGAAAAAGGAATCTATCAAACTTCTAACTTTAGAGTCCAACCAA...   \n",
       "\n",
       "                                     nucleotide_desc     lineage who_name  \\\n",
       "0  OA965538.1 Severe acute respiratory syndrome c...        B.39      NaN   \n",
       "1  OL454750.1 Severe acute respiratory syndrome c...       B.1.2      NaN   \n",
       "2  OW680892.1 Severe acute respiratory syndrome c...       B.1.1      NaN   \n",
       "3  OL444812.1 Severe acute respiratory syndrome c...           B      NaN   \n",
       "4  OL444814.1 Severe acute respiratory syndrome c...  B.1.177.12      NaN   \n",
       "\n",
       "  comment mapped_virus_name  \n",
       "0     NaN               NaN  \n",
       "1     NaN               NaN  \n",
       "2     NaN               NaN  \n",
       "3     NaN               NaN  \n",
       "4     NaN               NaN  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import results after GISAID\n",
    "nucleotide_variants_df = pd.read_csv(\"translated_sequences_variant.csv\")\n",
    "print(nucleotide_variants_df.shape)\n",
    "nucleotide_variants_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "58d4a161",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>nucleotide_id</th>\n",
       "      <th>protein_sequence</th>\n",
       "      <th>nucleotide_sequence</th>\n",
       "      <th>lineage</th>\n",
       "      <th>who_name</th>\n",
       "      <th>mapped_virus_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7X2M_1</td>\n",
       "      <td>OP964904.1</td>\n",
       "      <td>TNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFK...</td>\n",
       "      <td>GTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACT...</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>Beta</td>\n",
       "      <td>SARS-CoV2_Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7X2M_1</td>\n",
       "      <td>OP964904.1</td>\n",
       "      <td>MGILPSPGMPALLSLVSLLSVLLMGCVAETGTRFPNITNLCPFGEV...</td>\n",
       "      <td>GTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACT...</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>Beta</td>\n",
       "      <td>SARS-CoV2_Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6ZDH_1</td>\n",
       "      <td>OY602403.3</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...</td>\n",
       "      <td>B.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6ZDH_1</td>\n",
       "      <td>OY602403.3</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...</td>\n",
       "      <td>B.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6ZDH_1</td>\n",
       "      <td>OY602403.3</td>\n",
       "      <td>QCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPFF...</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...</td>\n",
       "      <td>B.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pdb_id nucleotide_id                                   protein_sequence  \\\n",
       "0  7X2M_1    OP964904.1  TNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFK...   \n",
       "1  7X2M_1    OP964904.1  MGILPSPGMPALLSLVSLLSVLLMGCVAETGTRFPNITNLCPFGEV...   \n",
       "2  6ZDH_1    OY602403.3  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...   \n",
       "3  6ZDH_1    OY602403.3  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...   \n",
       "4  6ZDH_1    OY602403.3  QCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPFF...   \n",
       "\n",
       "                                 nucleotide_sequence  lineage who_name  \\\n",
       "0  GTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACT...  B.1.351     Beta   \n",
       "1  GTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACT...  B.1.351     Beta   \n",
       "2  NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...      B.3      NaN   \n",
       "3  NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...      B.3      NaN   \n",
       "4  NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...      B.3      NaN   \n",
       "\n",
       "  mapped_virus_name  \n",
       "0    SARS-CoV2_Beta  \n",
       "1    SARS-CoV2_Beta  \n",
       "2               NaN  \n",
       "3               NaN  \n",
       "4               NaN  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Join the the two dataframes together based on their tBLASTn ID\n",
    "merged_lineages_df = pd.merge(translated_virus_sequences,nucleotide_variants_df,how=\"outer\",on=\"id_tblastn\")\n",
    "merged_lineages_df = merged_lineages_df[['id_y','id_tblastn','virus','sequence_y','lineage','who_name','mapped_virus_name']]\n",
    "merged_lineages_df = merged_lineages_df.rename(columns={'id_y':'pdb_id','id_tblastn':'nucleotide_id','virus':'protein_sequence','sequence_y':'nucleotide_sequence'})\n",
    "merged_lineages_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0f78609d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>nucleotide_id</th>\n",
       "      <th>protein_sequence</th>\n",
       "      <th>nucleotide_sequence</th>\n",
       "      <th>lineage</th>\n",
       "      <th>who_name</th>\n",
       "      <th>mapped_virus_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7X2M_1</td>\n",
       "      <td>OP964904.1</td>\n",
       "      <td>TNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFK...</td>\n",
       "      <td>GTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACT...</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>Beta</td>\n",
       "      <td>SARS-CoV2_Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7X2M_1</td>\n",
       "      <td>OP964904.1</td>\n",
       "      <td>MGILPSPGMPALLSLVSLLSVLLMGCVAETGTRFPNITNLCPFGEV...</td>\n",
       "      <td>GTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACT...</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>Beta</td>\n",
       "      <td>SARS-CoV2_Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>7WS4_1</td>\n",
       "      <td>OW023282.1</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...</td>\n",
       "      <td>BA.1</td>\n",
       "      <td>Omicron</td>\n",
       "      <td>SARS-CoV2_OmicronBA1;SARS-CoV2_Omicron-BA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>7WS4_1</td>\n",
       "      <td>OW023282.1</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...</td>\n",
       "      <td>BA.1</td>\n",
       "      <td>Omicron</td>\n",
       "      <td>SARS-CoV2_OmicronBA1;SARS-CoV2_Omicron-BA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>7QTI_5</td>\n",
       "      <td>ON334068.1</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>TAATAAAGGAGCTGGTGGCCATAGTTACGGCGCCGATCTAAAGTCA...</td>\n",
       "      <td>BA.1.18</td>\n",
       "      <td>Omicron</td>\n",
       "      <td>SARS-CoV2_OmicronBA1;SARS-CoV2_Omicron-BA1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     pdb_id nucleotide_id                                   protein_sequence  \\\n",
       "0    7X2M_1    OP964904.1  TNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFK...   \n",
       "1    7X2M_1    OP964904.1  MGILPSPGMPALLSLVSLLSVLLMGCVAETGTRFPNITNLCPFGEV...   \n",
       "115  7WS4_1    OW023282.1  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...   \n",
       "116  7WS4_1    OW023282.1  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...   \n",
       "124  7QTI_5    ON334068.1  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...   \n",
       "\n",
       "                                   nucleotide_sequence  lineage who_name  \\\n",
       "0    GTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACT...  B.1.351     Beta   \n",
       "1    GTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACT...  B.1.351     Beta   \n",
       "115  NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...     BA.1  Omicron   \n",
       "116  NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...     BA.1  Omicron   \n",
       "124  TAATAAAGGAGCTGGTGGCCATAGTTACGGCGCCGATCTAAAGTCA...  BA.1.18  Omicron   \n",
       "\n",
       "                              mapped_virus_name  \n",
       "0                                SARS-CoV2_Beta  \n",
       "1                                SARS-CoV2_Beta  \n",
       "115  SARS-CoV2_OmicronBA1;SARS-CoV2_Omicron-BA1  \n",
       "116  SARS-CoV2_OmicronBA1;SARS-CoV2_Omicron-BA1  \n",
       "124  SARS-CoV2_OmicronBA1;SARS-CoV2_Omicron-BA1  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop empty columns with no WHO name\n",
    "merged_lineages_df_with_who_name = merged_lineages_df.dropna(subset = ['who_name'])\n",
    "merged_lineages_df_with_who_name.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-17-71bf06d3212c>:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  merged_lineages_df_with_who_name['mapped_virus_name'] = merged_lineages_df_with_who_name['mapped_virus_name'].str.split(';')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>nucleotide_id</th>\n",
       "      <th>protein_sequence</th>\n",
       "      <th>nucleotide_sequence</th>\n",
       "      <th>lineage</th>\n",
       "      <th>who_name</th>\n",
       "      <th>mapped_virus_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7X2M_1</td>\n",
       "      <td>OP964904.1</td>\n",
       "      <td>TNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFK...</td>\n",
       "      <td>GTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACT...</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>Beta</td>\n",
       "      <td>SARS-CoV2_Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7X2M_1</td>\n",
       "      <td>OP964904.1</td>\n",
       "      <td>MGILPSPGMPALLSLVSLLSVLLMGCVAETGTRFPNITNLCPFGEV...</td>\n",
       "      <td>GTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACT...</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>Beta</td>\n",
       "      <td>SARS-CoV2_Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>7WS4_1</td>\n",
       "      <td>OW023282.1</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...</td>\n",
       "      <td>BA.1</td>\n",
       "      <td>Omicron</td>\n",
       "      <td>SARS-CoV2_OmicronBA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>7WS4_1</td>\n",
       "      <td>OW023282.1</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...</td>\n",
       "      <td>BA.1</td>\n",
       "      <td>Omicron</td>\n",
       "      <td>SARS-CoV2_Omicron-BA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>7WS4_1</td>\n",
       "      <td>OW023282.1</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...</td>\n",
       "      <td>BA.1</td>\n",
       "      <td>Omicron</td>\n",
       "      <td>SARS-CoV2_OmicronBA1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     pdb_id nucleotide_id                                   protein_sequence  \\\n",
       "0    7X2M_1    OP964904.1  TNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFK...   \n",
       "1    7X2M_1    OP964904.1  MGILPSPGMPALLSLVSLLSVLLMGCVAETGTRFPNITNLCPFGEV...   \n",
       "115  7WS4_1    OW023282.1  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...   \n",
       "115  7WS4_1    OW023282.1  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...   \n",
       "116  7WS4_1    OW023282.1  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...   \n",
       "\n",
       "                                   nucleotide_sequence  lineage who_name  \\\n",
       "0    GTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACT...  B.1.351     Beta   \n",
       "1    GTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACT...  B.1.351     Beta   \n",
       "115  NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...     BA.1  Omicron   \n",
       "115  NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...     BA.1  Omicron   \n",
       "116  NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...     BA.1  Omicron   \n",
       "\n",
       "         mapped_virus_name  \n",
       "0           SARS-CoV2_Beta  \n",
       "1           SARS-CoV2_Beta  \n",
       "115   SARS-CoV2_OmicronBA1  \n",
       "115  SARS-CoV2_Omicron-BA1  \n",
       "116   SARS-CoV2_OmicronBA1  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Explode the dataframe based on the mapped name\n",
    "merged_lineages_df_with_who_name['mapped_virus_name'] = merged_lineages_df_with_who_name['mapped_virus_name'].str.split(';')\n",
    "merged_lineages_df_with_who_name = merged_lineages_df_with_who_name.explode('mapped_virus_name')\n",
    "merged_lineages_df_with_who_name.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7394add4",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_lineages_df_with_who_name.to_csv('protein_variant_mapping.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa01fb7d",
   "metadata": {},
   "source": [
    "### Step 7: Generate training dataset\n",
    "Map these protein sequences based on their variant names to generate a training dataset, also import the SARS-CoV-2 reference strain for SARS-CoV2_WT and add it to our dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a6cf9708",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BEFORE ADDING WT SHAPE: (133, 7)\n",
      "AFTER ADDING WT SHAPE: (134, 7)\n",
      "pdb_id                                                                  \n",
      "nucleotide_id                                                           \n",
      "protein_sequence       MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...\n",
      "nucleotide_sequence                                                     \n",
      "lineage                                                                 \n",
      "who_name                                                                \n",
      "mapped_virus_name                                           SARS-CoV2_WT\n",
      "Name: 133, dtype: object\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>nucleotide_id</th>\n",
       "      <th>protein_sequence</th>\n",
       "      <th>nucleotide_sequence</th>\n",
       "      <th>lineage</th>\n",
       "      <th>who_name</th>\n",
       "      <th>mapped_virus_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7X2M_1</td>\n",
       "      <td>OP964904.1</td>\n",
       "      <td>TNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFK...</td>\n",
       "      <td>GTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACT...</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>Beta</td>\n",
       "      <td>SARS-CoV2_Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7X2M_1</td>\n",
       "      <td>OP964904.1</td>\n",
       "      <td>MGILPSPGMPALLSLVSLLSVLLMGCVAETGTRFPNITNLCPFGEV...</td>\n",
       "      <td>GTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACT...</td>\n",
       "      <td>B.1.351</td>\n",
       "      <td>Beta</td>\n",
       "      <td>SARS-CoV2_Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7WS4_1</td>\n",
       "      <td>OW023282.1</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...</td>\n",
       "      <td>BA.1</td>\n",
       "      <td>Omicron</td>\n",
       "      <td>SARS-CoV2_OmicronBA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7WS4_1</td>\n",
       "      <td>OW023282.1</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...</td>\n",
       "      <td>BA.1</td>\n",
       "      <td>Omicron</td>\n",
       "      <td>SARS-CoV2_Omicron-BA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7WS4_1</td>\n",
       "      <td>OW023282.1</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...</td>\n",
       "      <td>BA.1</td>\n",
       "      <td>Omicron</td>\n",
       "      <td>SARS-CoV2_OmicronBA1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pdb_id nucleotide_id                                   protein_sequence  \\\n",
       "0  7X2M_1    OP964904.1  TNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFK...   \n",
       "1  7X2M_1    OP964904.1  MGILPSPGMPALLSLVSLLSVLLMGCVAETGTRFPNITNLCPFGEV...   \n",
       "2  7WS4_1    OW023282.1  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...   \n",
       "3  7WS4_1    OW023282.1  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...   \n",
       "4  7WS4_1    OW023282.1  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...   \n",
       "\n",
       "                                 nucleotide_sequence  lineage who_name  \\\n",
       "0  GTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACT...  B.1.351     Beta   \n",
       "1  GTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACT...  B.1.351     Beta   \n",
       "2  NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...     BA.1  Omicron   \n",
       "3  NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...     BA.1  Omicron   \n",
       "4  NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...     BA.1  Omicron   \n",
       "\n",
       "       mapped_virus_name  \n",
       "0         SARS-CoV2_Beta  \n",
       "1         SARS-CoV2_Beta  \n",
       "2   SARS-CoV2_OmicronBA1  \n",
       "3  SARS-CoV2_Omicron-BA1  \n",
       "4   SARS-CoV2_OmicronBA1  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import the spike protein variant mapping\n",
    "spike_protein_variant_mapping = pd.read_csv(\"protein_variant_mapping.csv\")\n",
    "print(\"BEFORE ADDING WT SHAPE: \" + str(spike_protein_variant_mapping.shape))\n",
    "\n",
    "# Import the SARS-CoV-2 reference strain for SARS-CoV2_WT and add it into the dataframe\n",
    "wildtype_protein_sequence = \"\"\n",
    "for seq_record in SeqIO.parse(\"covid19_ref_strain.fasta\", \"fasta\"):\n",
    "        wildtype_protein_sequence = str(seq_record.seq)\n",
    "        break\n",
    "spike_protein_variant_mapping.loc[len(spike_protein_variant_mapping.index)] = [\"\",\"\",wildtype_protein_sequence,\"\",\"\",\"\",\"SARS-CoV2_WT\"]\n",
    "print(\"AFTER ADDING WT SHAPE: \" + str(spike_protein_variant_mapping.shape))\n",
    "print(spike_protein_variant_mapping.iloc[len(spike_protein_variant_mapping.index)-1])\n",
    "spike_protein_variant_mapping.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cd60c255",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Ab or Nb</th>\n",
       "      <th>Binds to</th>\n",
       "      <th>Doesn't Bind to</th>\n",
       "      <th>Neutralising Vs</th>\n",
       "      <th>Not Neutralising Vs</th>\n",
       "      <th>Protein + Epitope</th>\n",
       "      <th>Origin</th>\n",
       "      <th>VHorVHH</th>\n",
       "      <th>VL</th>\n",
       "      <th>...</th>\n",
       "      <th>Light J Gene</th>\n",
       "      <th>CDRH3</th>\n",
       "      <th>CDRL3</th>\n",
       "      <th>Structures</th>\n",
       "      <th>ABB Homology Model (if no structure)</th>\n",
       "      <th>Sources</th>\n",
       "      <th>Date Added</th>\n",
       "      <th>Last Updated</th>\n",
       "      <th>Update Description</th>\n",
       "      <th>Notes/Following Up?</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>036-2C2</td>\n",
       "      <td>Ab</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>S; RBD</td>\n",
       "      <td>B-cells; SARS-CoV2_WT Human Patient</td>\n",
       "      <td>QVQLVQSGAEVKKPGASVKVSCKASGYTFTGYYMHWVRQAPGQGLE...</td>\n",
       "      <td>DIQMTQSPSSLSASVGDRVTIICRASQYINSFLNWYQQKPGQAPKL...</td>\n",
       "      <td>...</td>\n",
       "      <td>IGKJ2 (Human)</td>\n",
       "      <td>ARGALRFLEWPILAY</td>\n",
       "      <td>QQSYSTPPYT</td>\n",
       "      <td>ND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanjeev Kumar et al., 2022 (https://www.scienc...</td>\n",
       "      <td>Dec 20, 2022</td>\n",
       "      <td>Dec 20, 2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>036-1B1</td>\n",
       "      <td>Ab</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>S; RBD</td>\n",
       "      <td>B-cells; SARS-CoV2_WT Human Patient</td>\n",
       "      <td>EGQLVESGGRLVQPGRSLRLSCAASGFTFDDYAMHWVRQAPGKGLE...</td>\n",
       "      <td>EIVLTQSPPTLSVSPGETATLSCRASLSLDRNLAWYQQKPGQAPRL...</td>\n",
       "      <td>...</td>\n",
       "      <td>IGKJ3 (Human)</td>\n",
       "      <td>VKDGNSDYGGNPI</td>\n",
       "      <td>QNYNKWPPLFT</td>\n",
       "      <td>ND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanjeev Kumar et al., 2022 (https://www.scienc...</td>\n",
       "      <td>Dec 20, 2022</td>\n",
       "      <td>Dec 20, 2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>036-1E2</td>\n",
       "      <td>Ab</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>S; RBD</td>\n",
       "      <td>B-cells; SARS-CoV2_WT Human Patient</td>\n",
       "      <td>QVQLVESGGGVVQPGRSLRLSCAAPGLSFRNYGMHWVRQAPGKGLE...</td>\n",
       "      <td>SYVLTQPPSVSVSPGQTARITCGGNNIGGKSVHWYQQKPGQAPVLI...</td>\n",
       "      <td>...</td>\n",
       "      <td>IGLJ2 (Human)</td>\n",
       "      <td>ARDRSGKDVLTGYPMFPAGMDV</td>\n",
       "      <td>QVWDVSSDHVV</td>\n",
       "      <td>ND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanjeev Kumar et al., 2022 (https://www.scienc...</td>\n",
       "      <td>Dec 20, 2022</td>\n",
       "      <td>Dec 20, 2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>036-1A10</td>\n",
       "      <td>Ab</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>S; RBD</td>\n",
       "      <td>B-cells; SARS-CoV2_WT Human Patient</td>\n",
       "      <td>QVQLVQSGAEVKKPGASVKVSCKASGYTFTSYGISWVRQAPGQGLE...</td>\n",
       "      <td>DIVMTQSPLSLPVTPGEPASISCRSSQSLLHSNGYNYLDWYLQKPG...</td>\n",
       "      <td>...</td>\n",
       "      <td>IGKJ2 (Human)</td>\n",
       "      <td>AREGYSYGTSWLGSSYYYYMDV</td>\n",
       "      <td>MQALQTPVRT</td>\n",
       "      <td>ND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanjeev Kumar et al., 2022 (https://www.scienc...</td>\n",
       "      <td>Dec 20, 2022</td>\n",
       "      <td>Dec 20, 2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>036-1F9</td>\n",
       "      <td>Ab</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>S; RBD</td>\n",
       "      <td>B-cells; SARS-CoV2_WT Human Patient</td>\n",
       "      <td>EVQVVESGGGLVQPGGSLRLSCAASGFPLSSYDMHWVRQRTGKGLE...</td>\n",
       "      <td>QSALTQPASVSGSPGQSITISCTGTSSDVGGFHYVSWYQQHPGKAP...</td>\n",
       "      <td>...</td>\n",
       "      <td>IGLJ3 (Human)</td>\n",
       "      <td>ARAQGYCGGGSCYSAYAFAI</td>\n",
       "      <td>SSYTSSSTWV</td>\n",
       "      <td>ND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanjeev Kumar et al., 2022 (https://www.scienc...</td>\n",
       "      <td>Dec 20, 2022</td>\n",
       "      <td>Dec 20, 2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Name Ab or Nb      Binds to Doesn't Bind to Neutralising Vs  \\\n",
       "11   036-2C2       Ab  SARS-CoV2_WT             NaN             NaN   \n",
       "12   036-1B1       Ab  SARS-CoV2_WT             NaN             NaN   \n",
       "14   036-1E2       Ab  SARS-CoV2_WT             NaN             NaN   \n",
       "16  036-1A10       Ab  SARS-CoV2_WT             NaN             NaN   \n",
       "22   036-1F9       Ab  SARS-CoV2_WT             NaN             NaN   \n",
       "\n",
       "   Not Neutralising Vs Protein + Epitope                               Origin  \\\n",
       "11        SARS-CoV2_WT            S; RBD  B-cells; SARS-CoV2_WT Human Patient   \n",
       "12        SARS-CoV2_WT            S; RBD  B-cells; SARS-CoV2_WT Human Patient   \n",
       "14        SARS-CoV2_WT            S; RBD  B-cells; SARS-CoV2_WT Human Patient   \n",
       "16        SARS-CoV2_WT            S; RBD  B-cells; SARS-CoV2_WT Human Patient   \n",
       "22        SARS-CoV2_WT            S; RBD  B-cells; SARS-CoV2_WT Human Patient   \n",
       "\n",
       "                                              VHorVHH  \\\n",
       "11  QVQLVQSGAEVKKPGASVKVSCKASGYTFTGYYMHWVRQAPGQGLE...   \n",
       "12  EGQLVESGGRLVQPGRSLRLSCAASGFTFDDYAMHWVRQAPGKGLE...   \n",
       "14  QVQLVESGGGVVQPGRSLRLSCAAPGLSFRNYGMHWVRQAPGKGLE...   \n",
       "16  QVQLVQSGAEVKKPGASVKVSCKASGYTFTSYGISWVRQAPGQGLE...   \n",
       "22  EVQVVESGGGLVQPGGSLRLSCAASGFPLSSYDMHWVRQRTGKGLE...   \n",
       "\n",
       "                                                   VL  ...   Light J Gene  \\\n",
       "11  DIQMTQSPSSLSASVGDRVTIICRASQYINSFLNWYQQKPGQAPKL...  ...  IGKJ2 (Human)   \n",
       "12  EIVLTQSPPTLSVSPGETATLSCRASLSLDRNLAWYQQKPGQAPRL...  ...  IGKJ3 (Human)   \n",
       "14  SYVLTQPPSVSVSPGQTARITCGGNNIGGKSVHWYQQKPGQAPVLI...  ...  IGLJ2 (Human)   \n",
       "16  DIVMTQSPLSLPVTPGEPASISCRSSQSLLHSNGYNYLDWYLQKPG...  ...  IGKJ2 (Human)   \n",
       "22  QSALTQPASVSGSPGQSITISCTGTSSDVGGFHYVSWYQQHPGKAP...  ...  IGLJ3 (Human)   \n",
       "\n",
       "                     CDRH3        CDRL3 Structures  \\\n",
       "11         ARGALRFLEWPILAY   QQSYSTPPYT         ND   \n",
       "12           VKDGNSDYGGNPI  QNYNKWPPLFT         ND   \n",
       "14  ARDRSGKDVLTGYPMFPAGMDV  QVWDVSSDHVV         ND   \n",
       "16  AREGYSYGTSWLGSSYYYYMDV   MQALQTPVRT         ND   \n",
       "22    ARAQGYCGGGSCYSAYAFAI   SSYTSSSTWV         ND   \n",
       "\n",
       "   ABB Homology Model (if no structure)  \\\n",
       "11                                  NaN   \n",
       "12                                  NaN   \n",
       "14                                  NaN   \n",
       "16                                  NaN   \n",
       "22                                  NaN   \n",
       "\n",
       "                                              Sources    Date Added  \\\n",
       "11  Sanjeev Kumar et al., 2022 (https://www.scienc...  Dec 20, 2022   \n",
       "12  Sanjeev Kumar et al., 2022 (https://www.scienc...  Dec 20, 2022   \n",
       "14  Sanjeev Kumar et al., 2022 (https://www.scienc...  Dec 20, 2022   \n",
       "16  Sanjeev Kumar et al., 2022 (https://www.scienc...  Dec 20, 2022   \n",
       "22  Sanjeev Kumar et al., 2022 (https://www.scienc...  Dec 20, 2022   \n",
       "\n",
       "    Last Updated Update Description Notes/Following Up?  \n",
       "11  Dec 20, 2022                NaN            Complete  \n",
       "12  Dec 20, 2022                NaN            Complete  \n",
       "14  Dec 20, 2022                NaN            Complete  \n",
       "16  Dec 20, 2022                NaN            Complete  \n",
       "22  Dec 20, 2022                NaN            Complete  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import the original CoVAbDb dataset\n",
    "covabdab_filepath = \"data/CoV-AbDab_201222.csv\"\n",
    "covabdab_df = pd.read_csv(covabdab_filepath)\n",
    "covabdab_df['VHorVHH'] = covabdab_df['VHorVHH'].replace('ND',np.nan,regex=False)\n",
    "covabdab_df['VL'] = covabdab_df['VL'].replace('ND',np.nan,regex=False)\n",
    "covabdab_df['Not Neutralising Vs'] = covabdab_df['Not Neutralising Vs'].replace('ND',np.nan)\n",
    "covabdab_df['Neutralising Vs'] = covabdab_df['Neutralising Vs'].replace('ND',np.nan)\n",
    "covabdab_df = covabdab_df.dropna(subset=['Neutralising Vs', 'Not Neutralising Vs'], how='all')\n",
    "covabdab_df = covabdab_df.dropna(subset=['VHorVHH', 'VL'], how='all')\n",
    "\n",
    "covabdab_df.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "144e26ff",
   "metadata": {},
   "source": [
    "We note that in the CovAbDb dataset, there are some differing names used to refer to the same variants in the `neutralizing vs` and `not neutralizing vs` columns based on the earlier work done to obtain the unique virus values. We will standardize the names as such:\n",
    "- All labels with prefix of \"SARS_CoV2_\" --> \"SARS-CoV2_\"\n",
    "- \"SARS-CoV2_WTSARS-CoV1\" --> \"SARS-CoV2_WT;SARS-CoV1\"\n",
    "- \"SARS-CoV2_BetaSARS-CoV2_Gamma\" --> \"SARS-CoV2_Beta;SARS-CoV2_Gamma\"\n",
    "- \"SRAS-CoV2_Beta\" --> \"SARS-CoV2_Beta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8e6bcc50",
   "metadata": {},
   "outputs": [],
   "source": [
    "def standardize_names(df, column):\n",
    "    df[column] = df[column].str.replace('SARS_CoV2_','SARS-CoV2_',regex=False)\n",
    "    df[column] = df[column].str.replace('SARS-CoV2_WTSARS-CoV1','SARS-CoV2_WT;SARS-CoV1',regex=False)\n",
    "    df[column] = df[column].str.replace('SARS-CoV2_BetaSARS-CoV2_Gamma','SARS-CoV2_Beta;SARS-CoV2_Gamma',regex=False)\n",
    "    df[column] = df[column].str.replace('SRAS-CoV2_Beta','SARS-CoV2_Beta',regex=False)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1286ca3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "covabdab_df_normalized = covabdab_df.copy()\n",
    "covabdab_df_normalized = standardize_names(covabdab_df_normalized, \"Binds to\")\n",
    "covabdab_df_normalized = standardize_names(covabdab_df_normalized, \"Doesn't Bind to\")\n",
    "covabdab_df_normalized = standardize_names(covabdab_df_normalized, \"Neutralising Vs\")\n",
    "covabdab_df_normalized = standardize_names(covabdab_df_normalized, \"Not Neutralising Vs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d419e372",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert values into lists for explosion\n",
    "covabdab_df_normalized['Neutralising Vs'] = covabdab_df_normalized['Neutralising Vs'].str.split(';')\n",
    "covabdab_df_normalized['Not Neutralising Vs'] = covabdab_df_normalized['Not Neutralising Vs'].str.split(';')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3eda018c",
   "metadata": {},
   "source": [
    "We now explode the data for both the neutralizing and not neutralising vs and not neutralising vs to form our dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f5df90e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>heavy_chain</th>\n",
       "      <th>light_chain</th>\n",
       "      <th>cdrh3</th>\n",
       "      <th>cdrl3</th>\n",
       "      <th>virus_type</th>\n",
       "      <th>virus_sequence</th>\n",
       "      <th>neutralising</th>\n",
       "      <th>weak_neutralisation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EVQLVESGGGLAQPGRSLRLSCAASGFTFDDYAMHWVRQAPGKGLE...</td>\n",
       "      <td>QSALTQPRSVSGSPGQSVTISCTGTSSDVGGYNYVSWYQQHPGKAP...</td>\n",
       "      <td>AKAEVPGYGSGWYQGFAS</td>\n",
       "      <td>CSYAGSYTGL</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EVQLVESGGGLIQPGGSLRLSCAASGITVSSNYMSWVRQAPGKGLE...</td>\n",
       "      <td>AIQLTQSPSSLSASVGDRVTITCRASQGISTYLAWYQQKPGKAPKL...</td>\n",
       "      <td>ARDLDYYGMDV</td>\n",
       "      <td>QQVNSYPPIT</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EVQLVESGGGLVQPGGSLRLSCAASGFTVSSHYMSWVRQAPGKGLE...</td>\n",
       "      <td>AIQLTQSPSSLSASVGDRVTITCRASQGISSYLAWYQQKPGKAPKL...</td>\n",
       "      <td>ARDSSWGPGYYGLDV</td>\n",
       "      <td>QQLNSLFT</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>QVQLVQSGAEVKKPGSSVKVSCKASGGTFSSYTITWVRQAPGQGLE...</td>\n",
       "      <td>QSLLTQPPSVSGAPGQRVTISCTGSNSNIGAGYDVHWYQQLPGTAP...</td>\n",
       "      <td>ARERGYSSSSSAWYFDL</td>\n",
       "      <td>QSYDSSLTGSL</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>QVQLVESGGGVVQPGRSLRLSCAASGFTFSNFAMYWVRQAPGKGLE...</td>\n",
       "      <td>SYELTQPPSVSVSPGQTARITCSGDALPKQYAYWYQKKPGQAPVLV...</td>\n",
       "      <td>ARDLEGEQWLLRDDYYYYYGMDV</td>\n",
       "      <td>QSADSSGTYRV</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         heavy_chain  \\\n",
       "0  EVQLVESGGGLAQPGRSLRLSCAASGFTFDDYAMHWVRQAPGKGLE...   \n",
       "1  EVQLVESGGGLIQPGGSLRLSCAASGITVSSNYMSWVRQAPGKGLE...   \n",
       "2  EVQLVESGGGLVQPGGSLRLSCAASGFTVSSHYMSWVRQAPGKGLE...   \n",
       "3  QVQLVQSGAEVKKPGSSVKVSCKASGGTFSSYTITWVRQAPGQGLE...   \n",
       "4  QVQLVESGGGVVQPGRSLRLSCAASGFTFSNFAMYWVRQAPGKGLE...   \n",
       "\n",
       "                                         light_chain                    cdrh3  \\\n",
       "0  QSALTQPRSVSGSPGQSVTISCTGTSSDVGGYNYVSWYQQHPGKAP...       AKAEVPGYGSGWYQGFAS   \n",
       "1  AIQLTQSPSSLSASVGDRVTITCRASQGISTYLAWYQQKPGKAPKL...              ARDLDYYGMDV   \n",
       "2  AIQLTQSPSSLSASVGDRVTITCRASQGISSYLAWYQQKPGKAPKL...          ARDSSWGPGYYGLDV   \n",
       "3  QSLLTQPPSVSGAPGQRVTISCTGSNSNIGAGYDVHWYQQLPGTAP...        ARERGYSSSSSAWYFDL   \n",
       "4  SYELTQPPSVSVSPGQTARITCSGDALPKQYAYWYQKKPGQAPVLV...  ARDLEGEQWLLRDDYYYYYGMDV   \n",
       "\n",
       "         cdrl3    virus_type  \\\n",
       "0   CSYAGSYTGL  SARS-CoV2_WT   \n",
       "1   QQVNSYPPIT  SARS-CoV2_WT   \n",
       "2     QQLNSLFT  SARS-CoV2_WT   \n",
       "3  QSYDSSLTGSL  SARS-CoV2_WT   \n",
       "4  QSADSSGTYRV  SARS-CoV2_WT   \n",
       "\n",
       "                                      virus_sequence  neutralising  \\\n",
       "0  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...          True   \n",
       "1  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...          True   \n",
       "2  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...          True   \n",
       "3  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...          True   \n",
       "4  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...          True   \n",
       "\n",
       "   weak_neutralisation  \n",
       "0                 True  \n",
       "1                False  \n",
       "2                 True  \n",
       "3                 True  \n",
       "4                 True  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Explode for neutralizing dataset\n",
    "neutralizing_dataset = covabdab_df_normalized.copy()\n",
    "neutralizing_dataset = neutralizing_dataset.dropna(subset=['Neutralising Vs'], how='all')\n",
    "neutralizing_dataset = neutralizing_dataset.explode('Neutralising Vs')\n",
    "# Preserve the weak neutralizing proeprty\n",
    "neutralizing_dataset['weak_neutralisation'] = neutralizing_dataset['Neutralising Vs'].str.contains('(weak)',case=False,na=False,regex=False)\n",
    "neutralizing_dataset['neutralising'] = True\n",
    "# Drop the (weak) notaion in the column for mapping later\n",
    "neutralizing_dataset['Neutralising Vs'] = neutralizing_dataset['Neutralising Vs'].str.replace(' (weak)',\"\",case=False,regex=False)\n",
    "neutralizing_dataset['Neutralising Vs'] = neutralizing_dataset['Neutralising Vs'].str.replace('(weak)',\"\",case=False,regex=False)\n",
    "# Rename columns\n",
    "neutralizing_dataset = neutralizing_dataset.rename(columns={'Neutralising Vs':'virus_type','VHorVHH':'heavy_chain','VL':'light_chain','CDRH3':'cdrh3','CDRL3':'cdrl3'})\n",
    "# Merge the virus spike protein with the dataset\n",
    "neutralizing_dataset = neutralizing_dataset.merge(spike_protein_variant_mapping,left_on=\"virus_type\",right_on=\"mapped_virus_name\")\n",
    "neutralizing_dataset = neutralizing_dataset.rename(columns={'protein_sequence':'virus_sequence'})\n",
    "# Retain columns of interest only -- we retain the CDRH3 and CDRL3 information as it may be of interest to us\n",
    "neutralizing_dataset = neutralizing_dataset[['heavy_chain','light_chain','cdrh3','cdrl3','virus_type','virus_sequence','neutralising','weak_neutralisation']]\n",
    "\n",
    "neutralizing_dataset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2b040014",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>heavy_chain</th>\n",
       "      <th>light_chain</th>\n",
       "      <th>cdrh3</th>\n",
       "      <th>cdrl3</th>\n",
       "      <th>virus_type</th>\n",
       "      <th>virus_sequence</th>\n",
       "      <th>neutralising</th>\n",
       "      <th>weak_neutralisation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>QVQLVQSGAEVKKPGASVKVSCKASGYTFTGYYMHWVRQAPGQGLE...</td>\n",
       "      <td>DIQMTQSPSSLSASVGDRVTIICRASQYINSFLNWYQQKPGQAPKL...</td>\n",
       "      <td>ARGALRFLEWPILAY</td>\n",
       "      <td>QQSYSTPPYT</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EGQLVESGGRLVQPGRSLRLSCAASGFTFDDYAMHWVRQAPGKGLE...</td>\n",
       "      <td>EIVLTQSPPTLSVSPGETATLSCRASLSLDRNLAWYQQKPGQAPRL...</td>\n",
       "      <td>VKDGNSDYGGNPI</td>\n",
       "      <td>QNYNKWPPLFT</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>QVQLVESGGGVVQPGRSLRLSCAAPGLSFRNYGMHWVRQAPGKGLE...</td>\n",
       "      <td>SYVLTQPPSVSVSPGQTARITCGGNNIGGKSVHWYQQKPGQAPVLI...</td>\n",
       "      <td>ARDRSGKDVLTGYPMFPAGMDV</td>\n",
       "      <td>QVWDVSSDHVV</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>QVQLVQSGAEVKKPGASVKVSCKASGYTFTSYGISWVRQAPGQGLE...</td>\n",
       "      <td>DIVMTQSPLSLPVTPGEPASISCRSSQSLLHSNGYNYLDWYLQKPG...</td>\n",
       "      <td>AREGYSYGTSWLGSSYYYYMDV</td>\n",
       "      <td>MQALQTPVRT</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EVQVVESGGGLVQPGGSLRLSCAASGFPLSSYDMHWVRQRTGKGLE...</td>\n",
       "      <td>QSALTQPASVSGSPGQSITISCTGTSSDVGGFHYVSWYQQHPGKAP...</td>\n",
       "      <td>ARAQGYCGGGSCYSAYAFAI</td>\n",
       "      <td>SSYTSSSTWV</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         heavy_chain  \\\n",
       "0  QVQLVQSGAEVKKPGASVKVSCKASGYTFTGYYMHWVRQAPGQGLE...   \n",
       "1  EGQLVESGGRLVQPGRSLRLSCAASGFTFDDYAMHWVRQAPGKGLE...   \n",
       "2  QVQLVESGGGVVQPGRSLRLSCAAPGLSFRNYGMHWVRQAPGKGLE...   \n",
       "3  QVQLVQSGAEVKKPGASVKVSCKASGYTFTSYGISWVRQAPGQGLE...   \n",
       "4  EVQVVESGGGLVQPGGSLRLSCAASGFPLSSYDMHWVRQRTGKGLE...   \n",
       "\n",
       "                                         light_chain                   cdrh3  \\\n",
       "0  DIQMTQSPSSLSASVGDRVTIICRASQYINSFLNWYQQKPGQAPKL...         ARGALRFLEWPILAY   \n",
       "1  EIVLTQSPPTLSVSPGETATLSCRASLSLDRNLAWYQQKPGQAPRL...           VKDGNSDYGGNPI   \n",
       "2  SYVLTQPPSVSVSPGQTARITCGGNNIGGKSVHWYQQKPGQAPVLI...  ARDRSGKDVLTGYPMFPAGMDV   \n",
       "3  DIVMTQSPLSLPVTPGEPASISCRSSQSLLHSNGYNYLDWYLQKPG...  AREGYSYGTSWLGSSYYYYMDV   \n",
       "4  QSALTQPASVSGSPGQSITISCTGTSSDVGGFHYVSWYQQHPGKAP...    ARAQGYCGGGSCYSAYAFAI   \n",
       "\n",
       "         cdrl3    virus_type  \\\n",
       "0   QQSYSTPPYT  SARS-CoV2_WT   \n",
       "1  QNYNKWPPLFT  SARS-CoV2_WT   \n",
       "2  QVWDVSSDHVV  SARS-CoV2_WT   \n",
       "3   MQALQTPVRT  SARS-CoV2_WT   \n",
       "4   SSYTSSSTWV  SARS-CoV2_WT   \n",
       "\n",
       "                                      virus_sequence  neutralising  \\\n",
       "0  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...         False   \n",
       "1  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...         False   \n",
       "2  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...         False   \n",
       "3  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...         False   \n",
       "4  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...         False   \n",
       "\n",
       "   weak_neutralisation  \n",
       "0                False  \n",
       "1                False  \n",
       "2                False  \n",
       "3                False  \n",
       "4                False  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Explode for neutralizing dataset\n",
    "not_neutralizing_dataset = covabdab_df_normalized.copy()\n",
    "not_neutralizing_dataset = not_neutralizing_dataset.dropna(subset=['Not Neutralising Vs'], how='all')\n",
    "not_neutralizing_dataset = not_neutralizing_dataset.explode('Not Neutralising Vs')\n",
    "# Preserve the weak neutralizing proeprty -- for not neutralizing, we set these both as false\n",
    "not_neutralizing_dataset['weak_neutralisation'] = False\n",
    "not_neutralizing_dataset['neutralising'] = False\n",
    "# Drop the (weak) notaion in the column for mapping later\n",
    "not_neutralizing_dataset['Not Neutralising Vs'] = not_neutralizing_dataset['Not Neutralising Vs'].str.replace(' (weak)',\"\",case=False,regex=False)\n",
    "not_neutralizing_dataset['Not Neutralising Vs'] = not_neutralizing_dataset['Not Neutralising Vs'].str.replace('(weak)',\"\",case=False,regex=False)\n",
    "# Rename columns\n",
    "not_neutralizing_dataset = not_neutralizing_dataset.rename(columns={'Not Neutralising Vs':'virus_type','VHorVHH':'heavy_chain','VL':'light_chain','CDRH3':'cdrh3','CDRL3':'cdrl3'})\n",
    "# Merge the virus spike protein with the dataset\n",
    "not_neutralizing_dataset = not_neutralizing_dataset.merge(spike_protein_variant_mapping,left_on=\"virus_type\",right_on=\"mapped_virus_name\")\n",
    "not_neutralizing_dataset = not_neutralizing_dataset.rename(columns={'protein_sequence':'virus_sequence'})\n",
    "# Retain columns of interest only -- we retain the CDRH3 and CDRL3 information as it may be of interest to us\n",
    "not_neutralizing_dataset = not_neutralizing_dataset[['heavy_chain','light_chain','cdrh3','cdrl3','virus_type','virus_sequence','neutralising','weak_neutralisation']]\n",
    "\n",
    "not_neutralizing_dataset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "71a6a1ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Exports for datasets\n",
    "neutralizing_dataset.to_csv(\"training_positive.csv\",index=False)\n",
    "not_neutralizing_dataset.to_csv(\"training_negative.csv\",index=False)\n",
    "# Concat both datasets together\n",
    "combined_dataset = pd.concat([neutralizing_dataset,not_neutralizing_dataset])\n",
    "combined_dataset.to_csv(\"training_combined.csv\",index=False)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
